// zent/lib/popover/Content.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _portal = require('../portal');

var _portal2 = _interopRequireDefault(_portal);

var _WindowResizeHandler = require('../utils/component/WindowResizeHandler');

var _WindowResizeHandler2 = _interopRequireDefault(_WindowResizeHandler);

var _WindowEventHandler = require('../utils/component/WindowEventHandler');

var _WindowEventHandler2 = _interopRequireDefault(_WindowEventHandler);

var _findPositionedParent = require('../utils/dom/findPositionedParent');

var _findPositionedParent2 = _interopRequireDefault(_findPositionedParent);

var _throttle = require('lodash/throttle');

var _throttle2 = _interopRequireDefault(_throttle);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _invisible = require('./placement/invisible');

var _invisible2 = _interopRequireDefault(_invisible);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function translateToContainerCoordinates(containerBB, bb) {
  var left = containerBB.left,
      top = containerBB.top;

  return {
    width: bb.width,
    height: bb.height,
    top: bb.top - top,
    left: bb.left - left,
    bottom: bb.bottom - top,
    right: bb.right - left
  };
}

/**
 * Like triggers, content can be replaced with your own implementation, all you have to do is extend this base class.
 *
 * The props on this class are all private.
 */

var PopoverContent = function (_ref) {
  (0, _inherits3['default'])(PopoverContent, _ref);

  function PopoverContent() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, PopoverContent);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = PopoverContent.__proto__ || Object.getPrototypeOf(PopoverContent)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      position: null
    }, _this.adjustPosition = function () {
      var content = _this.props.getContentNode();

      // 可能还未渲染出来，先放到一个不可见的位置
      if (!content) {
        _this.setState({
          position: (0, _invisible2['default'])(_this.props.prefix)
        });
        setTimeout(_this.adjustPosition, 0);

        return;
      }
      var contentBoundingBox = content.getBoundingClientRect();

      var anchor = _this.getAnchor();
      if (!anchor) {
        return;
      }
      var boundingBox = anchor.getBoundingClientRect();

      var parent = _this.getPositionedParent();
      if (!parent) {
        return;
      }
      var parentBoundingBox = parent.getBoundingClientRect();

      var relativeBB = translateToContainerCoordinates(parentBoundingBox, boundingBox);
      var relativeContainerBB = translateToContainerCoordinates(parentBoundingBox, parentBoundingBox);
      var position = _this.props.placement(_this.props.prefix, relativeBB, relativeContainerBB, {
        width: contentBoundingBox.width,
        height: contentBoundingBox.height
      }, {
        cushion: _this.props.cushion,
        anchor: anchor,
        container: parent,
        anchorBoundingBoxViewport: boundingBox,
        containerBoundingBoxViewport: parentBoundingBox
      });

      _this.setState({
        position: position
      }, _this.props.onPositionUpdated);
    }, _this.onWindowResize = (0, _throttle2['default'])(function (evt, delta) {
      if (_this.props.visible && (delta.deltaX !== 0 || delta.deltaY !== 0)) {
        _this.adjustPosition();
      }
    }, 16), _this.onWindowScroll = (0, _throttle2['default'])(_this.adjustPosition, 16), _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(PopoverContent, [{
    key: 'getAnchor',
    value: function getAnchor() {
      return this.props.getAnchor();
    }
  }, {
    key: 'getPositionedParent',
    value: function getPositionedParent() {
      // findPositionedParent returns null on fail
      if (this.positionedParent !== undefined) {
        return this.positionedParent;
      }

      var containerSelector = this.props.containerSelector;

      var container = document.querySelector(containerSelector);
      var parent = (0, _findPositionedParent2['default'])(container, true);
      this.positionedParent = parent;
      return parent;
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      var visible = this.props.visible;


      if (visible) {
        this.adjustPosition();
      }
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.visible && nextProps.visible !== this.props.visible) {
        this.adjustPosition();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          prefix = _props.prefix,
          className = _props.className,
          id = _props.id,
          visible = _props.visible,
          children = _props.children,
          containerSelector = _props.containerSelector;
      var position = this.state.position;


      if (!position) {
        return null;
      }

      var cls = (0, _classnames2['default'])(className, prefix + '-popover', id, position.toString());

      return _react2['default'].createElement(
        _portal2['default'],
        {
          prefix: prefix,
          visible: visible,
          selector: containerSelector,
          className: cls,
          css: position.getCSSStyle()
        },
        _react2['default'].createElement(
          'div',
          { className: prefix + '-popover-content' },
          children,
          _react2['default'].createElement(_WindowResizeHandler2['default'], { onResize: this.onWindowResize }),
          _react2['default'].createElement(_WindowEventHandler2['default'], {
            eventName: 'scroll',
            callback: this.onWindowScroll
          })
        )
      );
    }
  }]);
  return PopoverContent;
}(_react.PureComponent || _react.Component);

PopoverContent.propTypes = {
  children: _propTypes2['default'].node,

  prefix: _propTypes2['default'].string,

  id: _propTypes2['default'].string,

  getContentNode: _propTypes2['default'].func,

  visible: _propTypes2['default'].bool,

  // placement strategy
  placement: _propTypes2['default'].func,

  cushion: _propTypes2['default'].number,

  // A function that returns the anchor for this popover
  // () => Node
  getAnchor: _propTypes2['default'].func,

  // defaults to body
  containerSelector: _propTypes2['default'].string,

  onPositionUpdated: _propTypes2['default'].func
};
exports['default'] = PopoverContent;
