// zent/lib/datetimepicker/time/TimeCell.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var TimeCell = function (_ref) {
  (0, _inherits3['default'])(TimeCell, _ref);

  function TimeCell() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, TimeCell);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = TimeCell.__proto__ || Object.getPrototypeOf(TimeCell)).call.apply(_ref2, [this].concat(args))), _this), _this.onClickCell = function (cell) {
      !cell.isDisabled && _this.props.onSelect(cell.value);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(TimeCell, [{
    key: 'getTbody',
    value: function getTbody() {
      var _this2 = this;

      var cells = this.props.cells;

      return cells.map(function (row, i) {
        var tds = row.map(function (col, j) {
          return _react2['default'].createElement(
            'li',
            { className: 'grid-cell', role: 'gridcell', key: j },
            _react2['default'].createElement(
              'span',
              {
                onClick: function onClick() {
                  return _this2.onClickCell(col);
                },
                className: col.className,
                title: col.value
              },
              col.text
            )
          );
        });

        return _react2['default'].createElement(
          'ul',
          { className: 'panel-table__row', role: 'row', key: i },
          tds
        );
      });
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(
        'div',
        { className: 'panel-table__list' },
        this.getTbody()
      );
    }
  }]);
  return TimeCell;
}(_react.PureComponent || _react.Component);

TimeCell.propTypes = {
  onSelect: _propTypes2['default'].func
};
TimeCell.defaultProps = {
  onSelect: _constants.noop
};
exports['default'] = TimeCell;
