// zent/lib/design/preview/DesignPreviewItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _isFunction = require('lodash/isFunction');

var _isFunction2 = _interopRequireDefault(_isFunction);

var _scroll = require('../../utils/scroll');

var _scroll2 = _interopRequireDefault(_scroll);

var _offset = require('../utils/offset');

var _offset2 = _interopRequireDefault(_offset);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function scrollNodeToTop(node, offsets) {
  var pos = (0, _offset2['default'])(node);
  var top = (0, _isFunction2['default'])(offsets.top) ? offsets.top(pos.top) : pos.top + offsets.top;
  var left = (0, _isFunction2['default'])(offsets.left) ? offsets.left(pos.left) : pos.left + offsets.left;
  (0, _scroll2['default'])(document.body, left, top);
}

var DesignPreviewItem = function (_ref) {
  (0, _inherits3['default'])(DesignPreviewItem, _ref);

  function DesignPreviewItem() {
    (0, _classCallCheck3['default'])(this, DesignPreviewItem);
    return (0, _possibleConstructorReturn3['default'])(this, (DesignPreviewItem.__proto__ || Object.getPrototypeOf(DesignPreviewItem)).apply(this, arguments));
  }

  (0, _createClass3['default'])(DesignPreviewItem, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          prefix = _props.prefix;


      return _react2['default'].createElement(
        'div',
        { className: prefix + '-design-preview-item' },
        children
      );
    }
  }, {
    key: 'scrollTop',
    value: function scrollTop(offsets) {
      var node = (0, _reactDom.findDOMNode)(this);
      scrollNodeToTop(node, offsets);
    }
  }]);
  return DesignPreviewItem;
}(_react.PureComponent || _react.Component);

DesignPreviewItem.propTypes = {
  children: _propTypes2['default'].node.isRequired,
  prefix: _propTypes2['default'].string
};
DesignPreviewItem.defaultProps = {
  prefix: 'zent'
};
exports['default'] = DesignPreviewItem;
