// zent/lib/steps/components/Step.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _icon = require('../../icon');

var _icon2 = _interopRequireDefault(_icon);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Step = function (_ref) {
  (0, _inherits3['default'])(Step, _ref);

  function Step() {
    (0, _classCallCheck3['default'])(this, Step);
    return (0, _possibleConstructorReturn3['default'])(this, (Step.__proto__ || Object.getPrototypeOf(Step)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Step, [{
    key: 'render',
    value: function render() {
      var _classNames;

      var props = this.props;

      var prefix = props.prefix,
          isCurrentStep = props.isCurrentStep,
          _props$status = props.status,
          status = _props$status === undefined ? 'wait' : _props$status,
          isLastFinishStep = props.isLastFinishStep,
          stepLast = props.stepLast,
          stepNumber = props.stepNumber,
          title = props.title,
          description = props.description;


      var iconNode = void 0;

      if (status === 'finish') {
        iconNode = _react2['default'].createElement(_icon2['default'], { type: 'check-circle' });
      } else if (status === 'error') {
        iconNode = _react2['default'].createElement(_icon2['default'], { type: 'error-circle' });
      } else {
        iconNode = _react2['default'].createElement(
          'span',
          { className: prefix + '-icon' },
          stepNumber
        );
      }

      var classString = (0, _classnames2['default'])((_classNames = {}, (0, _defineProperty3['default'])(_classNames, prefix + '-steps-item', true), (0, _defineProperty3['default'])(_classNames, prefix + '-steps-status-finish', status === 'finish'), (0, _defineProperty3['default'])(_classNames, prefix + '-steps-status-error', status === 'error'), (0, _defineProperty3['default'])(_classNames, 'is-current', isCurrentStep), (0, _defineProperty3['default'])(_classNames, 'is-last-finish', isLastFinishStep), _classNames));

      return _react2['default'].createElement(
        'div',
        { className: classString },
        !stepLast && _react2['default'].createElement(
          'div',
          { className: prefix + '-steps-tail' },
          _react2['default'].createElement('i', null)
        ),
        _react2['default'].createElement(
          'div',
          { className: prefix + '-steps-step' },
          _react2['default'].createElement(
            'div',
            { className: prefix + '-step-head' },
            _react2['default'].createElement(
              'div',
              { className: prefix + '-step-head-inner' },
              iconNode
            )
          ),
          _react2['default'].createElement(
            'div',
            { className: prefix + '-step-main' },
            _react2['default'].createElement(
              'div',
              { className: prefix + '-step-title' },
              title
            ),
            description && _react2['default'].createElement(
              'div',
              { className: prefix + '-step-description' },
              description
            )
          )
        )
      );
    }
  }]);
  return Step;
}(_react.PureComponent || _react.Component);

Step.propTypes = {
  prefix: _propTypes2['default'].string,
  style: _propTypes2['default'].object,
  wrapperStyle: _propTypes2['default'].object,
  stepLast: _propTypes2['default'].bool,
  isCurrentStep: _propTypes2['default'].bool,
  isLastFinishStep: _propTypes2['default'].bool,
  stepNumber: _propTypes2['default'].string,
  status: _propTypes2['default'].string,
  title: _propTypes2['default'].node.isRequired,
  description: _propTypes2['default'].node
};
Step.defaultProps = {
  title: '',
  description: ''
};
exports['default'] = Step;
