// zent/lib/slider/InputField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _numberInput = require('../number-input');

var _numberInput2 = _interopRequireDefault(_numberInput);

var _pick = require('lodash/pick');

var _pick2 = _interopRequireDefault(_pick);

var _common = require('./common');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var InputField = function (_ref) {
  (0, _inherits3['default'])(InputField, _ref);

  function InputField() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, InputField);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = InputField.__proto__ || Object.getPrototypeOf(InputField)).call.apply(_ref2, [this].concat(args))), _this), _this.onChange = function (type, e) {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          value = _this$props.value;

      var newValue = Number(e.target.value);
      if (type === 'start') {
        newValue = [newValue, value[1]];
      } else if (type === 'end') {
        newValue = [value[0], newValue];
      }
      onChange && onChange(newValue);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(InputField, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          range = _props.range,
          value = _props.value,
          prefix = _props.prefix,
          restProps = (0, _objectWithoutProperties3['default'])(_props, ['range', 'value', 'prefix']);

      var numberInputProps = (0, _pick2['default'])(restProps, ['max', 'min', 'disabled', 'className', 'prefix']);
      return _react2['default'].createElement(
        'div',
        { className: prefix + '-slider-input' },
        range ? _react2['default'].createElement(
          'div',
          { className: prefix + '-slider-input' },
          _react2['default'].createElement(_numberInput2['default'], (0, _extends3['default'])({}, numberInputProps, {
            max: value[1],
            decimal: (0, _common.getDecimal)(restProps.step),
            onChange: this.onChange.bind(null, 'start'),
            value: value[0]
          })),
          _react2['default'].createElement(
            'span',
            { className: 'slider-input-line' },
            '-'
          ),
          _react2['default'].createElement(_numberInput2['default'], (0, _extends3['default'])({}, numberInputProps, {
            min: value[0],
            decimal: (0, _common.getDecimal)(restProps.step),
            onChange: this.onChange.bind(null, 'end'),
            value: value[1]
          }))
        ) : _react2['default'].createElement(_numberInput2['default'], (0, _extends3['default'])({}, numberInputProps, {
          decimal: (0, _common.getDecimal)(restProps.step),
          onChange: this.onChange.bind(null, 'single'),
          value: value
        }))
      );
    }
  }]);
  return InputField;
}(_react.PureComponent || _react.Component);

exports['default'] = InputField;
