// zent/lib/layout/Col.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Col = function (_ref) {
  (0, _inherits3['default'])(Col, _ref);

  function Col() {
    (0, _classCallCheck3['default'])(this, Col);
    return (0, _possibleConstructorReturn3['default'])(this, (Col.__proto__ || Object.getPrototypeOf(Col)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Col, [{
    key: 'render',
    value: function render() {
      var _cx;

      var _props = this.props,
          span = _props.span,
          offset = _props.offset,
          className = _props.className,
          prefix = _props.prefix,
          others = (0, _objectWithoutProperties3['default'])(_props, ['span', 'offset', 'className', 'prefix']);


      var classes = (0, _classnames2['default'])((_cx = {}, (0, _defineProperty3['default'])(_cx, prefix + '-col', true), (0, _defineProperty3['default'])(_cx, prefix + '-col-' + span, span), (0, _defineProperty3['default'])(_cx, prefix + '-col-offset-' + offset, offset), (0, _defineProperty3['default'])(_cx, className, className), _cx));

      return _react2['default'].createElement(
        'div',
        (0, _extends3['default'])({}, others, { className: classes }),
        this.props.children
      );
    }
  }]);
  return Col;
}(_react.PureComponent || _react.Component);

Col.propTypes = {
  span: _propTypes2['default'].number,
  offset: _propTypes2['default'].number,
  className: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string
};
Col.defaultProps = {
  prefix: 'zent'
};
exports['default'] = Col;
