// zent/lib/upload/components/UploadImageItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/**
 * 上传图片弹框
 */

var UploadImageItem = function (_ref) {
  (0, _inherits3['default'])(UploadImageItem, _ref);

  function UploadImageItem() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, UploadImageItem);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = UploadImageItem.__proto__ || Object.getPrototypeOf(UploadImageItem)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      hideDeleteIcon: false
    }, _this.handleRemove = function () {
      var _this$props = _this.props,
          index = _this$props.index,
          onDelete = _this$props.onDelete;

      onDelete(index);
    }, _this.handleDragStart = function () {
      _this.setState({
        hideDeleteIcon: true
      });
    }, _this.handleDragEnd = function () {
      _this.setState({
        hideDeleteIcon: false
      });
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(UploadImageItem, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          progress = _props.progress,
          src = _props.src,
          index = _props.index;
      var hideDeleteIcon = this.state.hideDeleteIcon;


      return _react2['default'].createElement(
        'li',
        {
          className: 'upload-local-image-item',
          onDragStart: this.handleDragStart,
          onDragEnd: this.handleDragEnd,
          'data-id': index
        },
        _react2['default'].createElement('div', {
          className: 'image-box',
          style: {
            backgroundImage: 'url(' + src + ')'
          }
        }),
        !hideDeleteIcon && _react2['default'].createElement(
          'span',
          { className: 'close-modal small', onClick: this.handleRemove },
          '\xD7'
        ),
        progress && _react2['default'].createElement(
          'div',
          { className: 'image-progress' },
          progress.toFixed(1) + '%'
        )
      );
    }
  }]);
  return UploadImageItem;
}(_react.PureComponent || _react.Component);

UploadImageItem.propTypes = {
  // 组件所在位置的下标
  index: _propTypes2['default'].number.isRequired,
  // 删除组件的回调函数
  onDelete: _propTypes2['default'].func.isRequired
};
exports['default'] = UploadImageItem;
