// zent/lib/datetimepicker/constants/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.commonPropTypes = exports.commonProps = exports.popPositionMap = exports.timeFnMap = exports.TIME_FORMAT = exports.noop = undefined;

var _noop = require('lodash/noop');

Object.defineProperty(exports, 'noop', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_noop)['default'];
  }
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _popover = require('../../popover');

var _popover2 = _interopRequireDefault(_popover);

var _noop2 = _interopRequireDefault(_noop);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var TIME_FORMAT = exports.TIME_FORMAT = 'HH:mm:ss';

var timeFnMap = exports.timeFnMap = {
  hour: 'setHours',
  minute: 'setMinutes',
  second: 'setSeconds'
};

var popPositionMap = exports.popPositionMap = {
  left: _popover2['default'].Position.AutoBottomLeft,
  right: _popover2['default'].Position.AutoBottomRight
};

var commonProps = exports.commonProps = {
  prefix: 'zent',
  confirmText: '',
  format: 'YYYY-MM-DD',
  popPosition: 'left',
  openPanel: false,
  disabledDate: _noop2['default'],
  onChange: _noop2['default'],
  isFooterVisble: false
};

var commonPropTypes = exports.commonPropTypes = {
  prefix: _propTypes2['default'].string,
  name: _propTypes2['default'].string,
  className: _propTypes2['default'].string,
  placeholder: _propTypes2['default'].string,
  confirmText: _propTypes2['default'].string,
  format: _propTypes2['default'].string,
  defaultTime: _propTypes2['default'].string,
  // onChange 返回值类型, date | number | string， 默认 string
  valueType: _propTypes2['default'].oneOf(['date', 'number', 'string']),
  popPosition: _propTypes2['default'].string,
  // min 和 max 可以传入和 format 一致的字符串或者 Date 实例
  min: _propTypes2['default'].oneOfType([_propTypes2['default'].string, _propTypes2['default'].number, _propTypes2['default'].instanceOf(Date)]),
  max: _propTypes2['default'].oneOfType([_propTypes2['default'].string, _propTypes2['default'].number, _propTypes2['default'].instanceOf(Date)]),
  disabledDate: _propTypes2['default'].func,
  onChange: _propTypes2['default'].func,
  onClick: _propTypes2['default'].func,
  onOpen: _propTypes2['default'].func,
  onClose: _propTypes2['default'].func
};
