// zent/lib/block-header/BlockHeader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _pop = require('../pop');

var _pop2 = _interopRequireDefault(_pop);

var _icon = require('../icon');

var _icon2 = _interopRequireDefault(_icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var BlockHeader = function (_Component) {
  (0, _inherits3['default'])(BlockHeader, _Component);

  function BlockHeader() {
    (0, _classCallCheck3['default'])(this, BlockHeader);
    return (0, _possibleConstructorReturn3['default'])(this, (BlockHeader.__proto__ || Object.getPrototypeOf(BlockHeader)).apply(this, arguments));
  }

  (0, _createClass3['default'])(BlockHeader, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          prefix = _props.prefix,
          content = _props.content,
          title = _props.title,
          tooltip = _props.tooltip,
          position = _props.position,
          className = _props.className,
          children = _props.children;

      return _react2['default'].createElement(
        'div',
        { className: (0, _classnames2['default'])(prefix + '-block-header', className) },
        title && _react2['default'].createElement(
          'div',
          { className: prefix + '-block-header__left' },
          _react2['default'].createElement(
            'h3',
            null,
            title
          )
        ),
        _react2['default'].createElement(
          'div',
          { className: prefix + '-block-header__pop' },
          tooltip && _react2['default'].createElement(
            _pop2['default'],
            {
              trigger: 'hover',
              centerArrow: true,
              position: position,
              content: _react2['default'].createElement(
                'p',
                { className: prefix + '-block-header__tooltip' },
                tooltip
              ),
              wrapperClassName: prefix + '-block-header__tooltip-trigger'
            },
            _react2['default'].createElement(_icon2['default'], { type: 'help-circle' })
          )
        ),
        _react2['default'].createElement(
          'div',
          { className: prefix + '-block-header__content' },
          content && content,
          children && children
        )
      );
    }
  }]);
  return BlockHeader;
}(_react.Component);

BlockHeader.propTypes = {
  className: _propTypes2['default'].string,
  title: _propTypes2['default'].string.isRequired,
  tooltip: _propTypes2['default'].node,
  content: _propTypes2['default'].node,
  position: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string
};
BlockHeader.defaultProps = {
  prefix: 'zent',
  className: '',
  position: 'top-right',
  tooltip: '',
  content: ''
};
exports['default'] = BlockHeader;
