// zent/lib/design/components/richtext/editor/plugins/Emotion.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports['default'] = function (options) {
  openDialog({
    dialogId: dialogId,
    className: 'zent-design-component-richtext__emotion',
    children: _react2['default'].createElement(Emotion, { callback: options.callback })
  });
};

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _dialog = require('../../../../../dialog');

var _dialog2 = _interopRequireDefault(_dialog);

var _tabs = require('../../../../../tabs');

var _tabs2 = _interopRequireDefault(_tabs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var TabPanel = _tabs2['default'].TabPanel;
var openDialog = _dialog2['default'].openDialog,
    closeDialog = _dialog2['default'].closeDialog;

var dialogId = 'zent_design_component_richtext_emotion_dialog';

var defaultTab = [{
  name: '精选',
  whichTpl: 'normal',
  tplInfo: 'Choice',
  wrapClass: 'jd',
  sum: 84,
  imgFolder: 'jx2/',
  imgName: 'j_00'
}, {
  name: '兔斯基',
  whichTpl: 'normal',
  tplInfo: 'Tuzki',
  wrapClass: 'tsj',
  sum: 40,
  imgFolder: 'tsj/',
  imgName: 't_00'
}, {
  name: '绿豆蛙',
  whichTpl: 'normal',
  tplInfo: 'Lvdouwa',
  wrapClass: 'ldw',
  sum: 52,
  imgFolder: 'ldw/',
  imgName: 'w_00'
}, {
  name: '波波',
  whichTpl: 'normal',
  tplInfo: 'Bobo',
  wrapClass: 'bb',
  sum: 63,
  imgFolder: 'bobo/',
  imgName: 'b_00'
}, {
  name: '北鼻猫',
  whichTpl: 'normal',
  tplInfo: 'BabyCat',
  wrapClass: 'cat',
  sum: 20,
  imgFolder: 'babycat/',
  imgName: 'C_00'
}, {
  name: '泡泡',
  whichTpl: 'pp',
  tplInfo: 'Bubble',
  wrapClass: 'pp',
  sum: 50,
  imgFolder: 'face/',
  imgName: 'i_f_'
}, {
  name: '有啊',
  whichTpl: 'normal',
  tplInfo: 'Youa',
  wrapClass: 'youa',
  sum: 44,
  imgFolder: 'youa/',
  imgName: 'y_00'
}, {
  name: 'QQ',
  whichTpl: 'qq',
  tplInfo: 'Qq',
  wrapClass: 'qq',
  sum: 105,
  imgFolder: 'qq/',
  imgName: ''
}];

var Emotion = function (_Component) {
  (0, _inherits3['default'])(Emotion, _Component);

  function Emotion(props) {
    (0, _classCallCheck3['default'])(this, Emotion);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (Emotion.__proto__ || Object.getPrototypeOf(Emotion)).call(this, props));

    _this.onTabChange = function (id) {
      _this.setState({
        activeId: id
      });
    };

    _this.state = {
      activeId: 'Choice'
    };
    return _this;
  }

  (0, _createClass3['default'])(Emotion, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.initTabs();
    }
  }, {
    key: 'initTabs',
    value: function initTabs() {
      var _this2 = this;

      this.allTabs = defaultTab.map(function (item, index) {
        var tabContent = [];

        for (var i = 1, len = item.sum; i <= len; i++) {
          var imgSrc = '' + item.imgFolder + item.imgName + (i < 10 ? '0' + i : i) + '.gif';
          var styleObj = {};
          var pos = '';

          if (item.whichTpl === 'pp') {
            styleObj = { backgroundPosition: 'left ' + (i - 1) * -25 + 'px' };
            pos = i % 12 <= 6 ? 'posRight' : 'posLeft';
          } else if (item.whichTpl === 'qq') {
            styleObj = { backgroundPosition: (i - 1) * -24 + 'px 50%' };
            pos = i % 15 <= 6 ? 'posRight' : 'posLeft';
          } else {
            styleObj = { backgroundPosition: 'left ' + (i - 1) * -35 + 'px' };
            pos = i % 12 <= 6 ? 'posRight' : 'posLeft';
          }

          tabContent.push(_react2['default'].createElement(
            'div',
            { key: item.tplInfo + '-' + i },
            _react2['default'].createElement('span', {
              className: item.wrapClass,
              style: styleObj,
              onClick: _this2.onClick.bind(_this2, item, imgSrc, pos)
            })
          ));
        }

        return _react2['default'].createElement(
          TabPanel,
          { tab: item.name, id: item.tplInfo, key: index },
          _react2['default'].createElement(
            'div',
            { className: 'emotion-content' },
            _react2['default'].createElement(
              'div',
              { className: 'emotion-list-wrapper' },
              tabContent
            )
          )
        );
      });
    }
  }, {
    key: 'onClick',
    value: function onClick(item, imgSrc) {
      var imgAddr = 'https://b.yzcdn.cn/v2';
      // App的WebView不支持//需要指明真正的协议
      imgAddr = imgAddr.replace(/^\/\//, 'https://');
      imgSrc = imgAddr + '/image/ueditor_emotion/' + imgSrc;

      closeDialog(dialogId);
      this.props.callback(imgSrc);
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(
        _tabs2['default'],
        {
          activeId: this.state.activeId,
          onTabChange: this.onTabChange,
          className: 'emotion-tabs'
        },
        this.allTabs
      );
    }
  }]);
  return Emotion;
}(_react.Component);
