// zent/lib/design/components/richtext/editor/plugins/Video.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports['default'] = function (options) {
  var _this2 = this;

  var closeDialog = _sweetalert2['default'].confirm({
    className: 'zent-design-component-richtext__video',
    title: '视频插入视频',
    content: _react2['default'].createElement(VideoForm, {
      callback: options.callback,
      onClose: function onClose() {
        return closeDialog();
      },
      ref: function ref(form) {
        return _this2.form = form;
      }
    }),
    onConfirm: function onConfirm() {
      _this2.form.getWrappedForm().saveVideo();
    }
  });
};

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _form = require('../../../../../form');

var _form2 = _interopRequireDefault(_form);

var _sweetalert = require('../../../../../sweetalert');

var _sweetalert2 = _interopRequireDefault(_sweetalert);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var createForm = _form2['default'].createForm,
    Field = _form2['default'].Field,
    InputField = _form2['default'].InputField;


var YOUKU_IMAGE = 'http://img.yzcdn.cn/public_files/2015/09/10/04eeb56eb29cbfbe29d67042be4d21ed.jpg';
var TUDOU_IMAGE = 'http://img.yzcdn.cn/public_files/2015/09/10/1640ba3f20b22d4b35a62d72831e8110.jpg';

var VideoForm = createForm({})(function (_Component) {
  (0, _inherits3['default'])(_class, _Component);

  function _class(props) {
    (0, _classCallCheck3['default'])(this, _class);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (_class.__proto__ || Object.getPrototypeOf(_class)).call(this, props));

    _this.preview = _this.preview.bind(_this);
    _this.width = props.width || 620;
    _this.height = props.height || _this.width * 0.75;

    // 自定义需要生成图片的width和height
    _this.imgWidth = props.imgWidth || 300;
    _this.imgHeight = props.imgHeight || 225;

    _this.iframeUrl = '';

    _this.state = {
      videoUrl: ''
    };
    return _this;
  }

  (0, _createClass3['default'])(_class, [{
    key: 'preview',
    value: function preview(e) {
      var videoUrl = e.target.value;
      if (e.type === 'paste') {
        setTimeout(this.previewVideo.bind(this, videoUrl), 1);
      } else {
        this.previewVideo(videoUrl);
      }
      this.setState({
        videoUrl: videoUrl
      });
    }
  }, {
    key: 'previewVideo',
    value: function previewVideo(url) {
      this.iframeUrl = this.processUrl(url);
      this.renderIframe();
    }
  }, {
    key: 'processUrl',
    value: function processUrl(url) {
      if (!url) {
        return;
      }

      var id = void 0;
      var iframeUrl = void 0;

      if (url.indexOf('v.qq.com') >= 0) {
        id = url.match(/vid=([^&]*)($|&)/);
        if (id) {
          iframeUrl = 'https://v.qq.com/iframe/player.html?vid=' + id[1] + '&tiny=0&auto=0';
        } else {
          id = url.match(/\/([0-9a-zA-Z]+).html/);
          if (id) {
            iframeUrl = 'https://v.qq.com/iframe/player.html?vid=' + id[1] + '&tiny=0&auto=0';
          }
        }
        if (!id) {
          return;
        }
      } else if (url.indexOf('v.youku.com') >= 0) {
        id = url.match(/id_(.*)\.html/);
        iframeUrl = 'http://player.youku.com/embed/' + id[1];
      } else if (url.indexOf('//player.youku.com/embed/') >= 0) {
        iframeUrl = url.match(/src="([^"]*)"/)[1];
      } else if (url.indexOf('tudou.com') >= 0) {
        id = url.match(/\/([\w\-]*)\.html/)[1];
        iframeUrl = 'http://www.tudou.com/programs/view/html5embed.action?code=' + id;
      } else {
        return;
      }

      return iframeUrl;
    }
  }, {
    key: 'renderIframe',
    value: function renderIframe(src) {
      src = src || this.iframeUrl;

      if (src) {
        var video = '<iframe src="' + src + '" width="' + this.width + '" height="' + this.height + '" allowfullscreen="true"></ifame>';

        if (this.isYouku() || this.isTudou()) {
          video = '<img src="' + this.getSiteLogoImage() + '" />';
        }

        document.getElementById('preview').innerHTML = video;
      } else {
        document.getElementById('preview').innerHTML = '<span>请复制腾讯、优酷视频地址输入框。</span>';
      }
    }
  }, {
    key: 'isYouku',
    value: function isYouku() {
      return this.iframeUrl && this.iframeUrl.match('youku');
    }
  }, {
    key: 'isTudou',
    value: function isTudou() {
      return this.iframeUrl && this.iframeUrl.match('tudou');
    }
  }, {
    key: 'getSiteLogoImage',
    value: function getSiteLogoImage() {
      if (this.isYouku()) {
        return YOUKU_IMAGE;
      }
      if (this.isTudou()) {
        return TUDOU_IMAGE;
      }
    }
  }, {
    key: 'saveVideo',
    value: function saveVideo() {
      if (!this.iframeUrl) return;

      var imgInfo = {
        url: this.iframeUrl,
        width: this.imgWidth,
        height: this.imgHeight
      };

      if (this.isYouku() || this.isTudou()) {
        imgInfo.html = '<a class="video-link" target="_blank" href="' + this.iframeUrl + '"><img src="' + this.getSiteLogoImage() + '" /></a>';
      }

      this.props.callback(imgInfo);
      this.props.onClose();
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(
        _form2['default'],
        {
          horizontal: true,
          className: 'video-content',
          onSubmit: this.props.handleSubmit(this.saveVideo)
        },
        _react2['default'].createElement(
          'strong',
          null,
          '\u4E3A\u4E86\u5728\u5FAE\u4FE1\u4E2D\u6709\u66F4\u597D\u7684\u4F53\u9A8C\uFF0C\u63A8\u8350\u4F7F\u7528',
          _react2['default'].createElement(
            'a',
            {
              href: 'http://v.qq.com',
              target: '_blank',
              rel: 'noopener noreferrer'
            },
            '\u817E\u8BAF\u89C6\u9891'
          ),
          '\u3002'
        ),
        _react2['default'].createElement(Field, {
          name: 'videoUrl',
          label: '\u89C6\u9891\u5730\u5740',
          placeholder: '\u590D\u5236\u89C6\u9891\u5730\u5740\u5230\u8FD9\u91CC',
          component: InputField,
          value: this.state.videoUrl,
          onInput: this.preview,
          onChange: this.preview
        }),
        _react2['default'].createElement(
          'div',
          { id: 'preview' },
          _react2['default'].createElement(
            'span',
            null,
            '\u9884\u89C8\u533A'
          )
        )
      );
    }
  }]);
  return _class;
}(_react.Component));
