// zent/lib/checkbox/Group.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function findIndex(array, predicate) {
  for (var i = 0; i < array.length; i++) {
    if (predicate(array[i])) {
      return i;
    }
  }

  return -1;
}

var Group = function (_ref) {
  (0, _inherits3['default'])(Group, _ref);

  function Group() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Group);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = Group.__proto__ || Object.getPrototypeOf(Group)).call.apply(_ref2, [this].concat(args))), _this), _this.onCheckboxChange = function (e) {
      var changedValue = e.target.value;
      var groupValue = _this.props.value.slice();
      var isValueEqual = _this.props.isValueEqual;

      var index = findIndex(groupValue, function (val) {
        return isValueEqual(val, changedValue);
      });

      if (index !== -1) {
        groupValue.splice(index, 1);
      } else {
        groupValue.push(changedValue);
      }

      _this.props.onChange(groupValue);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Group, [{
    key: 'render',
    value: function render() {
      var _this2 = this,
          _classNames;

      var _props = this.props,
          className = _props.className,
          prefix = _props.prefix,
          style = _props.style,
          isValueEqual = _props.isValueEqual,
          value = _props.value;

      var children = _react2['default'].Children.map(this.props.children, function (checkbox) {
        if (checkbox && checkbox.props) {
          return _react2['default'].cloneElement(checkbox, (0, _extends3['default'])({}, checkbox.props, {
            onChange: _this2.onCheckboxChange,
            checked: findIndex(value, function (val) {
              return isValueEqual(val, checkbox.props.value);
            }) !== -1,
            disabled: checkbox.props.disabled !== undefined ? checkbox.props.disabled : _this2.props.disabled,
            readOnly: checkbox.props.readOnly !== undefined ? checkbox.props.readOnly : _this2.props.readOnly
          }));
        }
      });

      var classString = (0, _classnames2['default'])((_classNames = {}, (0, _defineProperty3['default'])(_classNames, prefix + '-checkbox-group', true), (0, _defineProperty3['default'])(_classNames, className, !!className), _classNames));

      return _react2['default'].createElement(
        'div',
        { className: classString, style: style },
        children
      );
    }
  }]);
  return Group;
}(_react.PureComponent || _react.Component);

Group.propTypes = {
  value: _propTypes2['default'].array,
  isValueEqual: _propTypes2['default'].func,
  onChange: _propTypes2['default'].func,
  className: _propTypes2['default'].string,
  style: _propTypes2['default'].object,
  prefix: _propTypes2['default'].string
};
Group.defaultProps = {
  value: [],
  prefix: 'zent',
  className: '',
  style: {},
  onChange: function onChange() {},
  isValueEqual: function isValueEqual(a, b) {
    return a === b;
  }
};
exports['default'] = Group;
