// zent/lib/popover/placement/bottom-center.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _create = require('./create');

var _create2 = _interopRequireDefault(_create);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/**
 * ---------------
 * |   anchor     |
 * |              |
 * ---------------
 *    |popover|
 *    ---------
 */
function locate(anchorBoundingBox, containerBoundingBox, contentDimension, options) {
  var left = anchorBoundingBox.left,
      right = anchorBoundingBox.right,
      bottom = anchorBoundingBox.bottom;

  var middle = (left + right) / 2;
  var x = middle - contentDimension.width / 2;
  var y = bottom + options.cushion;

  return {
    getCSSStyle: function getCSSStyle() {
      return {
        position: 'absolute',
        left: Math.round(x) + 'px',
        top: Math.round(y) + 'px'
      };
    },


    name: 'position-bottom-center'
  };
}

var BottomCenter = (0, _create2['default'])(locate);

exports['default'] = BottomCenter;
