// zent/lib/popover/placement/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _create = require('./create');

var _create2 = _interopRequireDefault(_create);

var _bottomLeft = require('./bottom-left');

var _bottomLeft2 = _interopRequireDefault(_bottomLeft);

var _bottomCenter = require('./bottom-center');

var _bottomCenter2 = _interopRequireDefault(_bottomCenter);

var _bottomRight = require('./bottom-right');

var _bottomRight2 = _interopRequireDefault(_bottomRight);

var _leftTop = require('./left-top');

var _leftTop2 = _interopRequireDefault(_leftTop);

var _leftCenter = require('./left-center');

var _leftCenter2 = _interopRequireDefault(_leftCenter);

var _leftBottom = require('./left-bottom');

var _leftBottom2 = _interopRequireDefault(_leftBottom);

var _rightTop = require('./right-top');

var _rightTop2 = _interopRequireDefault(_rightTop);

var _rightCenter = require('./right-center');

var _rightCenter2 = _interopRequireDefault(_rightCenter);

var _rightBottom = require('./right-bottom');

var _rightBottom2 = _interopRequireDefault(_rightBottom);

var _topLeft = require('./top-left');

var _topLeft2 = _interopRequireDefault(_topLeft);

var _topCenter = require('./top-center');

var _topCenter2 = _interopRequireDefault(_topCenter);

var _topRight = require('./top-right');

var _topRight2 = _interopRequireDefault(_topRight);

var _autoBottomLeft = require('./auto-bottom-left');

var _autoBottomLeft2 = _interopRequireDefault(_autoBottomLeft);

var _autoBottomRight = require('./auto-bottom-right');

var _autoBottomRight2 = _interopRequireDefault(_autoBottomRight);

var _autoBottomCenter = require('./auto-bottom-center');

var _autoBottomCenter2 = _interopRequireDefault(_autoBottomCenter);

var _autoTopLeft = require('./auto-top-left');

var _autoTopLeft2 = _interopRequireDefault(_autoTopLeft);

var _autoTopRight = require('./auto-top-right');

var _autoTopRight2 = _interopRequireDefault(_autoTopRight);

var _autoTopCenter = require('./auto-top-center');

var _autoTopCenter2 = _interopRequireDefault(_autoTopCenter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

// Auto positions
exports['default'] = {
  create: _create2['default'],
  BottomLeft: _bottomLeft2['default'],
  BottomCenter: _bottomCenter2['default'],
  BottomRight: _bottomRight2['default'],
  LeftTop: _leftTop2['default'],
  LeftCenter: _leftCenter2['default'],
  LeftBottom: _leftBottom2['default'],
  RightTop: _rightTop2['default'],
  RightCenter: _rightCenter2['default'],
  RightBottom: _rightBottom2['default'],
  TopLeft: _topLeft2['default'],
  TopCenter: _topCenter2['default'],
  TopRight: _topRight2['default'],
  AutoBottomLeft: _autoBottomLeft2['default'],
  AutoBottomRight: _autoBottomRight2['default'],
  AutoBottomCenter: _autoBottomCenter2['default'],
  AutoTopLeft: _autoTopLeft2['default'],
  AutoTopRight: _autoTopRight2['default'],
  AutoTopCenter: _autoTopCenter2['default']
};
