// zent/lib/popover/placement/left-center.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _create = require('./create');

var _create2 = _interopRequireDefault(_create);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/**
 *           |--------------------|
 * ----------|                    |
 * | popover |   anchor           |
 * |---------|                    |
 *           |--------------------|
 */
function locate(anchorBoundingBox, containerBoundingBox, contentDimension, options) {
  var x = anchorBoundingBox.left - contentDimension.width - options.cushion;
  var middle = (anchorBoundingBox.top + anchorBoundingBox.bottom) / 2;
  var y = middle - contentDimension.height / 2;

  return {
    getCSSStyle: function getCSSStyle() {
      return {
        position: 'absolute',
        left: Math.round(x) + 'px',
        top: Math.round(y) + 'px'
      };
    },


    name: 'position-left-center'
  };
}

var LeftCenter = (0, _create2['default'])(locate);

exports['default'] = LeftCenter;
