// zent/lib/popover/placement/right-center.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _create = require('./create');

var _create2 = _interopRequireDefault(_create);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/**
* ----------------------
* |                    |----------
* |   anchor           | popover |
* |                    |---------
* |--------------------|
*/
function locate(anchorBoundingBox, containerBoundingBox, contentDimension, options) {
  var right = anchorBoundingBox.right,
      top = anchorBoundingBox.top,
      bottom = anchorBoundingBox.bottom;

  var x = right + options.cushion;
  var middle = (top + bottom) / 2;
  var y = middle - contentDimension.height / 2;

  return {
    getCSSStyle: function getCSSStyle() {
      return {
        position: 'absolute',
        left: Math.round(x) + 'px',
        top: Math.round(y) + 'px'
      };
    },


    name: 'position-right-center'
  };
}

var RightCenter = (0, _create2['default'])(locate);

exports['default'] = RightCenter;
