// zent/lib/popover/placement/top-center.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _create = require('./create');

var _create2 = _interopRequireDefault(_create);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/**
 *    ---------
 *    |popover|
 * ----------------
 * |   anchor     |
 * |              |
 * ----------------
 */
function locate(anchorBoundingBox, containerBoundingBox, contentDimension, options) {
  var right = anchorBoundingBox.right,
      left = anchorBoundingBox.left,
      top = anchorBoundingBox.top;

  var middle = (left + right) / 2;
  var x = middle - contentDimension.width / 2;
  var y = top - contentDimension.height - options.cushion;

  return {
    getCSSStyle: function getCSSStyle() {
      return {
        position: 'absolute',
        left: Math.round(x) + 'px',
        top: Math.round(y) + 'px'
      };
    },


    name: 'position-top-center'
  };
}

var TopCenter = (0, _create2['default'])(locate);

exports['default'] = TopCenter;
