// zent/lib/design/components/richtext/RichtextEditor.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _colorpicker = require('../../../colorpicker');

var _colorpicker2 = _interopRequireDefault(_colorpicker);

var _button = require('../../../button');

var _button2 = _interopRequireDefault(_button);

var _checkbox = require('../../../checkbox');

var _checkbox2 = _interopRequireDefault(_checkbox);

var _DesignEditor2 = require('../../editor/DesignEditor');

var _editor = require('./editor');

var _editor2 = _interopRequireDefault(_editor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var reserColor = '#f9f9f9';

var RichtextEditor = function (_DesignEditor) {
  (0, _inherits3['default'])(RichtextEditor, _DesignEditor);

  function RichtextEditor() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, RichtextEditor);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref = RichtextEditor.__proto__ || Object.getPrototypeOf(RichtextEditor)).call.apply(_ref, [this].concat(args))), _this), _this.handleResetBackground = function () {
      _this.onCustomInputChange('color')(reserColor);
    }, _this.onColorChange = function (color) {
      _this.onCustomInputChange('color')(color);
    }, _this.onFullscreenChange = function (e) {
      var isFullscreen = Number(e.target.checked);
      _this.onCustomInputChange('fullscreen')(isFullscreen);
    }, _this.onEditorChange = function (val) {
      _this.onCustomInputChange('content')(val);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(RichtextEditor, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          value = _props.value,
          richTextConfig = _props.richTextConfig;


      return _react2['default'].createElement(
        'div',
        { className: 'zent-design-component-richtext-editor' },
        _react2['default'].createElement(
          _DesignEditor2.ControlGroup,
          { focusOnLabelClick: false, label: '\u80CC\u666F\u989C\u8272\uFF1A' },
          _react2['default'].createElement(
            'div',
            { className: 'input-append' },
            _react2['default'].createElement(_colorpicker2['default'], {
              className: 'zent-design-component-richtext-editor__color-picker-popover',
              color: value.color,
              onChange: this.onColorChange
            }),
            _react2['default'].createElement(
              _button2['default'],
              { onClick: this.handleResetBackground },
              '\u91CD\u7F6E'
            )
          ),
          _react2['default'].createElement(
            'label',
            { htmlFor: 'fullscreen', className: 'control-label' },
            '\u662F\u5426\u5168\u5C4F\uFF1A'
          ),
          _react2['default'].createElement(
            _checkbox2['default'],
            {
              className: 'zent-design-component-richtext-editor-checkbox-wrap',
              name: 'fullscreen',
              checked: value.fullscreen,
              onChange: this.onFullscreenChange
            },
            '\u5168\u5C4F\u663E\u793A'
          )
        ),
        _react2['default'].createElement(
          'div',
          { className: 'zent-design-component-richtext-editor-group' },
          _react2['default'].createElement(_editor2['default'], (0, _extends3['default'])({
            value: value.content,
            onChange: this.onEditorChange,
            editorConfig: {
              initialFrameWidth: 386,
              initialFrameHeight: 600
            }
          }, richTextConfig))
        )
      );
    }
  }], [{
    key: 'getInitialValue',
    value: function getInitialValue() {
      return {
        color: reserColor,
        content: '',
        fullscreen: 0
      };
    }
  }]);
  return RichtextEditor;
}(_DesignEditor2.DesignEditor);

RichtextEditor.defaultProps = {
  richTextConfig: {}
};
RichtextEditor.designType = 'rich_text';
RichtextEditor.designDescription = '富文本';
exports['default'] = RichtextEditor;
