// zent/lib/portal/withESCToClose.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports['default'] = withESCToClose;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _omit = require('lodash/omit');

var _omit2 = _interopRequireDefault(_omit);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/*
  Not exported in index.js.

  Provides an HOC component for ESC to close functionality, useful in some cases.
*/
function withESCToClose(Closable) {
  var _class, _temp2;

  return _temp2 = _class = function (_ref) {
    (0, _inherits3['default'])(ESCToCloseWrapper, _ref);

    function ESCToCloseWrapper() {
      var _ref2;

      var _temp, _this, _ret;

      (0, _classCallCheck3['default'])(this, ESCToCloseWrapper);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = ESCToCloseWrapper.__proto__ || Object.getPrototypeOf(ESCToCloseWrapper)).call.apply(_ref2, [this].concat(args))), _this), _this.onKeyUp = function (evt) {
        if (evt.keyCode === 27) {
          _this.props.onClose();
        }
      }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
    }

    (0, _createClass3['default'])(ESCToCloseWrapper, [{
      key: 'on',
      value: function on() {
        document.body.addEventListener('keyup', this.onKeyUp, true);
      }
    }, {
      key: 'off',
      value: function off() {
        document.body.removeEventListener('keyup', this.onKeyUp, true);
      }
    }, {
      key: 'componentDidMount',
      value: function componentDidMount() {
        if (this.props.visible) {
          this.on();
        }
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        if (this.props.visible) {
          this.off();
        }
      }
    }, {
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(nextProps) {
        if (nextProps.visible !== this.props.visible) {
          if (nextProps.visible) {
            this.on();
          } else {
            this.off();
          }
        }
      }
    }, {
      key: 'render',
      value: function render() {
        var props = (0, _omit2['default'])(this.props, ['onClose']);
        return _react2['default'].createElement(Closable, props);
      }
    }]);
    return ESCToCloseWrapper;
  }(_react.PureComponent || _react.Component), _class.propTypes = {
    visible: _propTypes2['default'].bool.isRequired,
    onClose: _propTypes2['default'].func.isRequired
  }, _temp2;
}
