// zent/lib/datetimepicker/year/YearPanel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _PanelHeader = require('../common/PanelHeader');

var _PanelHeader2 = _interopRequireDefault(_PanelHeader);

var _YearPanelBody = require('./YearPanelBody');

var _YearPanelBody2 = _interopRequireDefault(_YearPanelBody);

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var YearPanel = function (_ref) {
  (0, _inherits3['default'])(YearPanel, _ref);

  function YearPanel() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, YearPanel);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = YearPanel.__proto__ || Object.getPrototypeOf(YearPanel)).call.apply(_ref2, [this].concat(args))), _this), _this.prevYears = function () {
      var _this$props = _this.props,
          actived = _this$props.actived,
          onChange = _this$props.onChange;

      var prev = actived.getFullYear() - 12;
      onChange(prev, true);
    }, _this.nextYears = function () {
      var _this$props2 = _this.props,
          actived = _this$props2.actived,
          onChange = _this$props2.onChange;

      var next = actived.getFullYear() + 12;
      onChange(next, true);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(YearPanel, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          actived = _props.actived,
          onSelect = _props.onSelect,
          selected = _props.selected,
          disabledDate = _props.disabledDate;

      var currentYear = parseInt(actived.getFullYear(), 10);
      var title = currentYear - 4 + '~' + (currentYear + 7);

      return _react2['default'].createElement(
        'div',
        { className: 'year-panel' },
        _react2['default'].createElement(_PanelHeader2['default'], {
          title: title,
          onClickTitle: _constants.noop,
          prev: this.prevYears,
          next: this.nextYears
        }),
        _react2['default'].createElement(_YearPanelBody2['default'], {
          actived: actived,
          selected: selected,
          disabledDate: disabledDate,
          onSelect: onSelect
        })
      );
    }
  }]);
  return YearPanel;
}(_react.PureComponent || _react.Component);

exports['default'] = YearPanel;
