// zent/lib/datetimepicker/year/YearPanelBody.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _utils = require('../utils/');

var _PanelCell = require('../common/PanelCell');

var _PanelCell2 = _interopRequireDefault(_PanelCell);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var ROW = 4;
var COL = 3;

var YearPanelBody = function (_ref) {
  (0, _inherits3['default'])(YearPanelBody, _ref);

  function YearPanelBody() {
    (0, _classCallCheck3['default'])(this, YearPanelBody);
    return (0, _possibleConstructorReturn3['default'])(this, (YearPanelBody.__proto__ || Object.getPrototypeOf(YearPanelBody)).apply(this, arguments));
  }

  (0, _createClass3['default'])(YearPanelBody, [{
    key: 'getYears',
    value: function getYears() {
      var years = [];
      var index = 0;
      var _props = this.props,
          actived = _props.actived,
          selected = _props.selected,
          disabledDate = _props.disabledDate;

      var beginYear = actived.getFullYear() - 4;

      for (var rowIndex = 0; rowIndex < ROW; rowIndex++) {
        years[rowIndex] = [];
        for (var colIndex = 0; colIndex < COL; colIndex++) {
          var val = beginYear + index;
          var isCurrent = val === _utils.CURRENT_YEAR;
          var isSelected = selected instanceof Date && selected.getFullYear() === val;
          var isDisabled = disabledDate && disabledDate(val);
          var className = (0, _classnames2['default'])({
            'year-panel__cell panel__cell': true,
            'panel__cell--current': isCurrent,
            'panel__cell--selected': isSelected,
            'panel__cell--disabled': isDisabled
          });
          years[rowIndex][colIndex] = {
            text: val,
            value: val,
            title: '' + val,
            className: className,
            isDisabled: isDisabled
          };
          index++;
        }
      }

      return years;
    }
  }, {
    key: 'render',
    value: function render() {
      var onSelect = this.props.onSelect;

      var years = this.getYears();

      return _react2['default'].createElement(
        'div',
        { className: 'year-table panel-table' },
        _react2['default'].createElement(_PanelCell2['default'], { onSelect: onSelect, cells: years })
      );
    }
  }]);
  return YearPanelBody;
}(_react.PureComponent || _react.Component);

YearPanelBody.propTypes = {
  actived: _propTypes2['default'].instanceOf(Date),
  onSelect: _propTypes2['default'].func
};
exports['default'] = YearPanelBody;
