// zent/lib/select/components/Option.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Option = function (_ref) {
  (0, _inherits3['default'])(Option, _ref);

  function Option(props) {
    (0, _classCallCheck3['default'])(this, Option);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (Option.__proto__ || Object.getPrototypeOf(Option)).call(this, props));

    _this.optionClickHandler = _this.optionClickHandler.bind(_this);
    return _this;
  }

  (0, _createClass3['default'])(Option, [{
    key: 'optionClickHandler',
    value: function optionClickHandler(ev) {
      this.props.onClick(ev, this.props.cid);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          text = _props.text,
          value = _props.value;

      return _react2['default'].createElement(
        'span',
        {
          value: value,
          className: className,
          onClick: this.optionClickHandler,
          onMouseEnter: this.props.onMouseEnter
        },
        text
      );
    }
  }]);
  return Option;
}(_react.PureComponent || _react.Component);

Option.propTypes = {
  prefixCls: _propTypes2['default'].string,
  cid: _propTypes2['default'].string,
  value: _propTypes2['default'].any,
  text: _propTypes2['default'].any,
  placeholder: _propTypes2['default'].string,
  onMouseEnter: _propTypes2['default'].func
};

exports['default'] = Option;
