// zent/lib/design/components/image-ad/ImageAdPreview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _isEmpty = require('lodash/isEmpty');

var _isEmpty2 = _interopRequireDefault(_isEmpty);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var ImageAdPreview = function (_Component) {
  (0, _inherits3['default'])(ImageAdPreview, _Component);

  function ImageAdPreview() {
    (0, _classCallCheck3['default'])(this, ImageAdPreview);
    return (0, _possibleConstructorReturn3['default'])(this, (ImageAdPreview.__proto__ || Object.getPrototypeOf(ImageAdPreview)).apply(this, arguments));
  }

  (0, _createClass3['default'])(ImageAdPreview, [{
    key: 'render',
    value: function render() {
      var _cx;

      var _props = this.props,
          value = _props.value,
          prefix = _props.prefix;
      var size = value.size,
          images = value.images;


      if ((0, _isEmpty2['default'])(images)) {
        return _react2['default'].createElement(
          'div',
          {
            className: (0, _classnames2['default'])(prefix + '-design-component-image-ad-preview', prefix + '-design-component-image-ad-preview--no-data')
          },
          '\u70B9\u51FB\u7F16\u8F91\u56FE\u7247\u5E7F\u544A'
        );
      }

      return _react2['default'].createElement(
        'div',
        {
          className: (0, _classnames2['default'])(prefix + '-design-component-image-ad-preview', (_cx = {}, (0, _defineProperty3['default'])(_cx, prefix + '-design-component-image-ad-preview--large', size === _constants.IMAGE_SIZE.LARGE), (0, _defineProperty3['default'])(_cx, prefix + '-design-component-image-ad-preview--small', size === _constants.IMAGE_SIZE.SMALL), _cx))
        },
        images.map(function (img) {
          var id = img[_constants.IMAGE_AD_ENTRY_UUID_KEY];
          // eslint-disable-next-line
          var url = img.linkUrl || 'javascript:void(0);';
          var title = img.linkTitle;

          return _react2['default'].createElement(
            'a',
            {
              key: id,
              className: prefix + '-design-component-image-ad-preview__image',
              href: url,
              target: '_blank',
              rel: 'noopener noreferrer'
            },
            _react2['default'].createElement(
              'div',
              {
                className: prefix + '-design-component-image-ad-preview__image-img'
              },
              _react2['default'].createElement('img', { src: img.imageUrl, alt: title }),
              title && _react2['default'].createElement(
                'div',
                {
                  className: prefix + '-design-component-image-ad-preview__image-title'
                },
                title
              )
            )
          );
        })
      );
    }
  }]);
  return ImageAdPreview;
}(_react.Component);

exports['default'] = ImageAdPreview;
