// zent/lib/design/components/image-ad/ImageAdEditor.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radio = require('../../../radio');

var _radio2 = _interopRequireDefault(_radio);

var _icon = require('../../../icon');

var _icon2 = _interopRequireDefault(_icon);

var _reactBeautifulDnd = require('react-beautiful-dnd');

var _createObjectURL = require('../../../utils/createObjectURL');

var _createObjectURL2 = _interopRequireDefault(_createObjectURL);

var _findIndex = require('lodash/findIndex');

var _findIndex2 = _interopRequireDefault(_findIndex);

var _isEmpty = require('lodash/isEmpty');

var _isEmpty2 = _interopRequireDefault(_isEmpty);

var _DesignEditor2 = require('../../editor/DesignEditor');

var _constants = require('./constants');

var _ImageEntry = require('./ImageEntry');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var RadioGroup = _radio2['default'].Group;

var ImageAdEditor = function (_DesignEditor) {
  (0, _inherits3['default'])(ImageAdEditor, _DesignEditor);

  function ImageAdEditor(props) {
    (0, _classCallCheck3['default'])(this, ImageAdEditor);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (ImageAdEditor.__proto__ || Object.getPrototypeOf(ImageAdEditor)).call(this, props));

    _this.onAddImageEntry = function (evt) {
      var files = evt.target.files;

      var imageUrl = (0, _createObjectURL2['default'])(files[0]);
      var _this$props = _this.props,
          value = _this$props.value,
          onChange = _this$props.onChange;


      onChange({
        images: value.images.concat((0, _ImageEntry.createEmptyImageEntry)({ imageUrl: imageUrl }))
      });
    };

    _this.removeImageEntry = function (id) {
      return function () {
        var _this$props2 = _this.props,
            images = _this$props2.value.images,
            onChange = _this$props2.onChange;


        onChange({
          images: images.filter(function (img) {
            return img[_constants.IMAGE_AD_ENTRY_UUID_KEY] !== id;
          })
        });
      };
    };

    _this.prependImageEntry = function (id) {
      return function () {
        var _this$props3 = _this.props,
            images = _this$props3.value.images,
            onChange = _this$props3.onChange;

        var index = (0, _findIndex2['default'])(images, function (img) {
          return img[_constants.IMAGE_AD_ENTRY_UUID_KEY] === id;
        });
        if (index !== -1) {
          var newImages = images.slice();
          newImages.splice(index, 0, (0, _ImageEntry.createEmptyImageEntry)());

          onChange({
            images: newImages
          });
        }
      };
    };

    _this.appendImageEntry = function (id) {
      return function () {
        var _this$props4 = _this.props,
            images = _this$props4.value.images,
            onChange = _this$props4.onChange;

        var index = (0, _findIndex2['default'])(images, function (img) {
          return img[_constants.IMAGE_AD_ENTRY_UUID_KEY] === id;
        });
        if (index !== -1) {
          var newImages = images.slice();
          newImages.splice(index + 1, 0, (0, _ImageEntry.createEmptyImageEntry)());

          onChange({
            images: newImages
          });
        }
      };
    };

    _this.onImageEntryChange = function (id) {
      return function (delta) {
        var _this$props5 = _this.props,
            images = _this$props5.value.images,
            onChange = _this$props5.onChange;

        onChange({
          images: images.map(function (img) {
            if (img[_constants.IMAGE_AD_ENTRY_UUID_KEY] !== id) {
              return img;
            }

            return (0, _extends3['default'])({}, img, delta);
          })
        });
      };
    };

    _this.state = (0, _extends3['default'])({}, _this.state, {
      localImage: ''
    });
    return _this;
  }

  (0, _createClass3['default'])(ImageAdEditor, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          prefix = _props.prefix,
          showError = _props.showError,
          validation = _props.validation,
          value = _props.value;
      var localImage = this.state.localImage;

      var imageErrors = validation.images;
      var allowAddImage = this.isAddImageEntryAllowed();

      return _react2['default'].createElement(
        'div',
        { className: prefix + '-design-component-image-ad-editor' },
        _react2['default'].createElement(
          _DesignEditor2.ControlGroup,
          {
            label: '\u663E\u793A\u5927\u5C0F:',
            showError: showError || this.getMetaProperty('size', 'touched'),
            error: validation.size
          },
          _react2['default'].createElement(
            RadioGroup,
            { value: value.size, onChange: this.onInputChange },
            _react2['default'].createElement(
              _radio2['default'],
              { name: 'size', value: _constants.IMAGE_SIZE.LARGE },
              '\u5927\u56FE'
            ),
            _react2['default'].createElement(
              _radio2['default'],
              { name: 'size', value: _constants.IMAGE_SIZE.SMALL },
              '\u5C0F\u56FE'
            )
          )
        ),
        _react2['default'].createElement(
          _reactBeautifulDnd.Droppable,
          {
            droppableId: prefix + '-design-component-image-ad-editor__entry-list',
            type: _constants.IMAGE_AD_DND_TYPE,
            direction: 'vertical'
          },
          function (provided, snapshot) {
            return _react2['default'].createElement(
              'ul',
              {
                ref: provided.innerRef,
                className: prefix + '-design-component-image-ad-editor__entry-list'
              },
              value.images.map(function (img) {
                var imageId = img[_constants.IMAGE_AD_ENTRY_UUID_KEY];

                return _react2['default'].createElement(
                  'li',
                  {
                    key: imageId,
                    className: prefix + '-design-component-image-ad-editor__entry'
                  },
                  _react2['default'].createElement(_ImageEntry.ImageEntry, {
                    prefix: prefix,
                    imageId: imageId,
                    imageUrl: img.imageUrl,
                    linkTitle: img.linkTitle,
                    linkUrl: img.linkUrl,
                    onChange: _this2.onImageEntryChange(imageId),
                    error: showError && imageErrors ? imageErrors[imageId] : ''
                  }),
                  !snapshot.isDraggingOver && _react2['default'].createElement(_icon2['default'], {
                    type: 'close-circle',
                    className: prefix + '-design-component-image-ad-editor__entry-close-btn',
                    onClick: _this2.removeImageEntry(imageId)
                  }),
                  !snapshot.isDraggingOver && allowAddImage && _react2['default'].createElement(_icon2['default'], {
                    type: 'plus',
                    className: prefix + '-design-component-image-ad-editor__entry-prepend-btn',
                    onClick: _this2.prependImageEntry(imageId)
                  }),
                  !snapshot.isDraggingOver && allowAddImage && _react2['default'].createElement(_icon2['default'], {
                    type: 'plus',
                    className: prefix + '-design-component-image-ad-editor__entry-append-btn',
                    onClick: _this2.appendImageEntry(imageId)
                  })
                );
              }),
              provided.placeholder
            );
          }
        ),
        allowAddImage && _react2['default'].createElement(
          'a',
          {
            className: prefix + '-design-component-image-ad-editor__add-entry-btn'
          },
          _react2['default'].createElement(
            'b',
            null,
            '+'
          ),
          '\u6DFB\u52A0\u4E00\u4E2A\u5E7F\u544A',
          _react2['default'].createElement('input', {
            type: 'file',
            accept: 'image/gif, image/jpeg, image/png',
            title: '',
            value: localImage,
            onChange: this.onAddImageEntry
          })
        ),
        _react2['default'].createElement(
          'div',
          { className: prefix + '-design-component-image-ad-editor__hint' },
          '\u6700\u591A\u6DFB\u52A0 ',
          _constants.IMAGE_AD_LIMIT,
          ' \u4E2A\u5E7F\u544A\uFF0C\u62D6\u52A8\u9009\u4E2D\u7684\u56FE\u7247\u5E7F\u544A\u53EF\u5BF9\u5176\u6392\u5E8F'
        )
      );
    }
  }, {
    key: 'isAddImageEntryAllowed',
    value: function isAddImageEntryAllowed() {
      var images = this.props.value.images;


      return images.length < _constants.IMAGE_AD_LIMIT;
    }
  }, {
    key: 'shouldHandleDragEnd',
    value: function shouldHandleDragEnd(type) {
      return type === _constants.IMAGE_AD_DND_TYPE;
    }
  }, {
    key: 'onDragEnd',
    value: function onDragEnd(result) {
      var source = result.source,
          destination = result.destination;

      // dropped outside

      if (!destination) {
        return;
      }

      var _props2 = this.props,
          value = _props2.value,
          onChange = _props2.onChange;

      var newValue = (0, _extends3['default'])({}, value, {
        images: this.reorder(value.images, source.index, destination.index)
      });

      onChange(newValue);
    }
  }], [{
    key: 'getInitialValue',
    value: function getInitialValue() {
      return {
        size: _constants.IMAGE_SIZE.SMALL,
        images: []
      };
    }
  }, {
    key: 'validate',
    value: function validate(value) {
      return new Promise(function (resolve) {
        var errors = {};

        errors.images = value.images.reduce(function (imageErrors, img) {
          if (!img.imageUrl) {
            imageErrors[img[_constants.IMAGE_AD_ENTRY_UUID_KEY]] = '请选择广告图片';
          }
          return imageErrors;
        }, {});

        // 如果没有错误就删除这个 key
        if ((0, _isEmpty2['default'])(errors.images)) {
          delete errors.images;
        }

        resolve(errors);
      });
    }
  }]);
  return ImageAdEditor;
}(_DesignEditor2.DesignEditor);

ImageAdEditor.designType = 'image-ad';
ImageAdEditor.designDescription = _react2['default'].createElement(
  'span',
  null,
  '\u56FE\u7247',
  _react2['default'].createElement('br', null),
  '\u5E7F\u544A'
);
exports['default'] = ImageAdEditor;
