// zent/lib/design/components/image-ad/ImageEntry.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageEntry = undefined;

var _extends3 = require('babel-runtime/helpers/extends');

var _extends4 = _interopRequireDefault(_extends3);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.createEmptyImageEntry = createEmptyImageEntry;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _input = require('../../../input');

var _input2 = _interopRequireDefault(_input);

var _uuid = require('../../../utils/uuid');

var _uuid2 = _interopRequireDefault(_uuid);

var _reactBeautifulDnd = require('react-beautiful-dnd');

var _createObjectURL = require('../../../utils/createObjectURL');

var _createObjectURL2 = _interopRequireDefault(_createObjectURL);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var ImageEntry = exports.ImageEntry = function (_Component) {
  (0, _inherits3['default'])(ImageEntry, _Component);

  function ImageEntry() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, ImageEntry);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref = ImageEntry.__proto__ || Object.getPrototypeOf(ImageEntry)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      localImage: ''
    }, _this.onTitleChange = function (evt) {
      var value = evt.target.value;

      _this.props.onChange({ linkTitle: value });
    }, _this.onUrlChange = function (evt) {
      var value = evt.target.value;

      _this.props.onChange({ linkUrl: value });
    }, _this.onImageChange = function (evt) {
      var files = evt.target.files;

      var imageUrl = (0, _createObjectURL2['default'])(files[0]);
      _this.props.onChange({ imageUrl: imageUrl });
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(ImageEntry, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          imageId = _props.imageId,
          imageUrl = _props.imageUrl,
          linkTitle = _props.linkTitle,
          linkUrl = _props.linkUrl,
          error = _props.error,
          prefix = _props.prefix;
      var localImage = this.state.localImage;


      return _react2['default'].createElement(
        _reactBeautifulDnd.Draggable,
        { draggableId: imageId, type: _constants.IMAGE_AD_DND_TYPE },
        function (provided) {
          var _cx;

          return _react2['default'].createElement(
            'div',
            null,
            _react2['default'].createElement(
              'div',
              (0, _extends4['default'])({
                className: prefix + '-design-component-image-ad-editor__image-entry',
                ref: provided.innerRef,
                style: provided.draggableStyle
              }, provided.dragHandleProps),
              _react2['default'].createElement(
                'div',
                {
                  className: prefix + '-design-component-image-ad-editor__image-entry-image-container'
                },
                _react2['default'].createElement('img', { src: imageUrl, alt: linkTitle }),
                _react2['default'].createElement(
                  'div',
                  {
                    className: (0, _classnames2['default'])(prefix + '-design-component-image-ad-editor__image-entry-image-upload', (_cx = {}, (0, _defineProperty3['default'])(_cx, prefix + '-design-component-image-ad-editor__image-entry-image-upload--has-image', imageUrl), (0, _defineProperty3['default'])(_cx, prefix + '-design-component-image-ad-editor__image-entry-image-upload--no-image', !imageUrl), _cx))
                  },
                  imageUrl ? _react2['default'].createElement(
                    'span',
                    null,
                    '\u91CD\u65B0\u4E0A\u4F20'
                  ) : _react2['default'].createElement(
                    'a',
                    null,
                    _react2['default'].createElement(
                      'b',
                      null,
                      '+'
                    ),
                    '\u6DFB\u52A0\u56FE\u7247'
                  ),
                  _react2['default'].createElement('input', {
                    type: 'file',
                    accept: 'image/gif, image/jpeg, image/png',
                    title: '',
                    value: localImage,
                    onChange: _this2.onImageChange
                  })
                )
              ),
              _react2['default'].createElement(
                'div',
                {
                  className: prefix + '-design-component-image-ad-editor__image-entry-controls'
                },
                _react2['default'].createElement(
                  'div',
                  {
                    className: prefix + '-design-component-image-ad-editor__image-entry-image-control'
                  },
                  _react2['default'].createElement(
                    'label',
                    null,
                    '\u6807\u9898:'
                  ),
                  _react2['default'].createElement(_input2['default'], { value: linkTitle, onChange: _this2.onTitleChange })
                ),
                _react2['default'].createElement(
                  'div',
                  {
                    className: prefix + '-design-component-image-ad-editor__image-entry-image-control'
                  },
                  _react2['default'].createElement(
                    'label',
                    null,
                    '\u94FE\u63A5:'
                  ),
                  _react2['default'].createElement(_input2['default'], { value: linkUrl, onChange: _this2.onUrlChange })
                )
              ),
              error && _react2['default'].createElement(
                'div',
                {
                  className: prefix + '-design-component-image-ad-editor__image-entry-error'
                },
                error
              )
            ),
            provided.placeholder
          );
        }
      );
    }
  }]);
  return ImageEntry;
}(_react.Component);

ImageEntry.propTypes = {
  prefix: _propTypes2['default'].string,
  imageId: _propTypes2['default'].string,
  imageUrl: _propTypes2['default'].string,
  linkUrl: _propTypes2['default'].string,
  linkTitle: _propTypes2['default'].string,
  error: _propTypes2['default'].string,
  onChange: _propTypes2['default'].func.isRequired
};
function createEmptyImageEntry(override) {
  return (0, _extends4['default'])((0, _defineProperty3['default'])({
    imageUrl: '',
    linkTitle: '',
    linkUrl: ''
  }, _constants.IMAGE_AD_ENTRY_UUID_KEY, (0, _uuid2['default'])()), override);
}
