// zent/lib/sortable/Sortable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _sortablejs = require('sortablejs');

var _sortablejs2 = _interopRequireDefault(_sortablejs);

var _isFunction = require('lodash/isFunction');

var _isFunction2 = _interopRequireDefault(_isFunction);

var _reorder = require('../utils/reorder');

var _reorder2 = _interopRequireDefault(_reorder);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Sortable = function (_ref) {
  (0, _inherits3['default'])(Sortable, _ref);

  function Sortable() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Sortable);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = Sortable.__proto__ || Object.getPrototypeOf(Sortable)).call.apply(_ref2, [this].concat(args))), _this), _this.initSortable = function (instance) {
      var _this$props = _this.props,
          prefix = _this$props.prefix,
          options = _this$props.options,
          onChange = _this$props.onChange,
          filterClass = _this$props.filterClass,
          children = _this$props.children,
          rest = (0, _objectWithoutProperties3['default'])(_this$props, ['prefix', 'options', 'onChange', 'filterClass', 'children']);


      if (!instance) {
        return;
      }

      var sortableOptions = (0, _extends3['default'])({
        filter: '.' + filterClass,
        ghostClass: prefix + '-ghost',
        chosenClass: prefix + '-chosen',
        dragClass: prefix + '-drag',
        fallbackClass: prefix + '-fallback',
        onMove: function onMove(e) {
          return e.related.className !== filterClass;
        },
        onEnd: function onEnd(e) {
          var items = _this.props.items;
          var oldIndex = e.oldIndex,
              newIndex = e.newIndex;

          var newItems = (0, _reorder2['default'])(items, oldIndex, newIndex);
          onChange && onChange(newItems);
        }
      }, rest);

      _this.sortable = _sortablejs2['default'].create(instance, sortableOptions);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Sortable, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.sortable) {
        this.sortable.destroy();
        this.sortable = null;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          prefix = _props.prefix,
          className = _props.className,
          children = _props.children,
          Com = _props.tag;

      var classString = (0, _classnames2['default'])(prefix + '-sortable', className);

      return _react2['default'].createElement(
        Com,
        {
          ref: function ref(instance) {
            return _this2.initSortable(instance);
          },
          className: classString
        },
        children
      );
    }
  }]);
  return Sortable;
}(_react.PureComponent || _react.Component);

Sortable.propTypes = {
  className: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string,
  items: _propTypes2['default'].array,
  onChange: function onChange(props, propName, componentName) {
    if (props.items && !props[propName] || !props.items && props[propName]) {
      throw new Error('prop items and onChange must appear at the same time.');
    }
    if (props[propName] && !(0, _isFunction2['default'])(props[propName])) {
      throw new Error('Invalid prop ' + propName + ' supplied to ' + componentName + ', expects a function.');
    }
  },

  tag: _propTypes2['default'].string,
  sort: _propTypes2['default'].bool,
  filterClass: _propTypes2['default'].string,
  group: _propTypes2['default'].oneOfType([_propTypes2['default'].string, _propTypes2['default'].object]),
  delay: _propTypes2['default'].number,
  animation: _propTypes2['default'].number,
  handle: _propTypes2['default'].string,
  ghostClass: _propTypes2['default'].string,
  chosenClass: _propTypes2['default'].string,
  dragClass: _propTypes2['default'].string,
  forceFallback: _propTypes2['default'].bool,
  fallbackClass: _propTypes2['default'].string,
  fallbackOnBody: _propTypes2['default'].bool,
  fallbackTolerance: _propTypes2['default'].number,
  scroll: _propTypes2['default'].bool,
  scrollFn: _propTypes2['default'].func,
  scrollSensitivity: _propTypes2['default'].number,
  scrollSpeed: _propTypes2['default'].number,
  setData: _propTypes2['default'].func,
  onStart: _propTypes2['default'].func,
  onEnd: _propTypes2['default'].func,
  onAdd: _propTypes2['default'].func,
  onUpdate: _propTypes2['default'].func,
  onSort: _propTypes2['default'].func,
  onRemove: _propTypes2['default'].func,
  onFilter: _propTypes2['default'].func,
  onMove: _propTypes2['default'].func,
  onClone: _propTypes2['default'].func
};
Sortable.defaultProps = {
  prefix: 'zent',
  tag: 'div'
};
exports['default'] = Sortable;
