// zent/lib/alert/Alert.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _isFunction = require('lodash/isFunction');

var _isFunction2 = _interopRequireDefault(_isFunction);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

// 忽略不支持的style
var styleClassMap = {
  info: 'alert-style-info',
  warning: 'alert-style-warning',

  // error as an alias to danger
  error: 'alert-style-danger',
  danger: 'alert-style-danger'
};

// 忽略不支持的size
var sizeClassMap = {
  normal: 'alert-size-normal',
  large: 'alert-size-large'
};

var Alert = function (_ref) {
  (0, _inherits3['default'])(Alert, _ref);

  function Alert() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Alert);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = Alert.__proto__ || Object.getPrototypeOf(Alert)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      closed: false
    }, _this.onClose = function () {
      _this.setState({
        closed: true
      }, function () {
        // onClose是在*关闭以后*被调用的
        var onClose = _this.props.onClose;

        if ((0, _isFunction2['default'])(onClose)) {
          onClose();
        }
      });
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Alert, [{
    key: 'render',
    value: function render() {
      var _cx;

      var closed = this.state.closed;

      if (closed) {
        return null;
      }

      var _props = this.props,
          type = _props.type,
          prefix = _props.prefix,
          rounded = _props.rounded,
          className = _props.className,
          closable = _props.closable,
          size = _props.size,
          children = _props.children;

      var containerCls = (0, _classnames2['default'])(prefix + '-alert', prefix + '-' + styleClassMap[type], prefix + '-' + sizeClassMap[size], (_cx = {}, (0, _defineProperty3['default'])(_cx, className, !!className), (0, _defineProperty3['default'])(_cx, prefix + '-alert-border-rounded', rounded), (0, _defineProperty3['default'])(_cx, prefix + '-alert-closable', closable), _cx));

      return _react2['default'].createElement(
        'div',
        { className: containerCls },
        closable && _react2['default'].createElement(
          'div',
          { className: prefix + '-alert-close-wrapper' },
          _react2['default'].createElement(
            'span',
            {
              className: prefix + '-alert-close-btn',
              onClick: this.onClose
            },
            '\xD7'
          )
        ),
        _react2['default'].createElement(
          'div',
          { className: prefix + '-alert-content-wrapper' },
          _react2['default'].createElement(
            'div',
            { className: prefix + '-alert-content' },
            children
          )
        )
      );
    }
  }]);
  return Alert;
}(_react.PureComponent || _react.Component);

Alert.propTypes = {
  type: _propTypes2['default'].oneOf(['info', 'warning', 'danger', 'error']).isRequired,
  size: _propTypes2['default'].oneOf(['large', 'normal']),
  rounded: _propTypes2['default'].bool,
  closable: _propTypes2['default'].bool,
  onClose: _propTypes2['default'].func,
  children: _propTypes2['default'].node,
  className: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string
};
Alert.defaultProps = {
  type: 'info',
  size: 'normal',
  closable: false,
  rounded: false,
  className: '',
  prefix: 'zent'
};
exports['default'] = Alert;
