// zent/lib/badge/Badge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Badge = function (_ref) {
  (0, _inherits3['default'])(Badge, _ref);

  function Badge() {
    (0, _classCallCheck3['default'])(this, Badge);
    return (0, _possibleConstructorReturn3['default'])(this, (Badge.__proto__ || Object.getPrototypeOf(Badge)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Badge, [{
    key: 'render',
    value: function render() {
      var _cx;

      var _props = this.props,
          count = _props.count,
          maxCount = _props.maxCount,
          dot = _props.dot,
          showZero = _props.showZero,
          className = _props.className,
          prefix = _props.prefix,
          children = _props.children;

      var containerCls = (0, _classnames2['default'])((_cx = {}, (0, _defineProperty3['default'])(_cx, prefix + '-badge', true), (0, _defineProperty3['default'])(_cx, prefix + '-badge-none-cont', !children), (0, _defineProperty3['default'])(_cx, className, !!className), _cx));

      var renderCount = function renderCount() {
        var countEle = null;
        if (dot) {
          countEle = _react2['default'].createElement('span', { className: prefix + '-badge-count ' + prefix + '-badge-dot' });
        } else if (count > 0 || count === 0 && showZero) {
          countEle = _react2['default'].createElement(
            'span',
            { className: prefix + '-badge-count' },
            count > maxCount ? maxCount + '+' : count
          );
        }
        return countEle;
      };

      return _react2['default'].createElement(
        'div',
        { className: containerCls },
        children ? _react2['default'].createElement(
          'div',
          { className: prefix + '-badge-content' },
          children
        ) : null,
        renderCount()
      );
    }
  }]);
  return Badge;
}(_react.PureComponent || _react.Component);

Badge.propTypes = {
  count: _propTypes2['default'].number,
  maxCount: _propTypes2['default'].number,
  dot: _propTypes2['default'].bool,
  showZero: _propTypes2['default'].bool,
  children: _propTypes2['default'].node,
  className: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string
};
Badge.defaultProps = {
  count: 0,
  maxCount: 99,
  dot: false,
  showZero: false,
  className: '',
  prefix: 'zent'
};
exports['default'] = Badge;
