// zent/lib/form/form-components/FormComponentField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormSwitchField = exports.FormSelectField = exports.FormRadioGroupField = exports.FormNumberInputField = exports.FormInputField = exports.FormDateRangePickerField = exports.FormColorPickerField = exports.FormCheckboxGroupField = exports.FormCheckboxField = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _CheckboxField = require('./CheckboxField');

var _CheckboxField2 = _interopRequireDefault(_CheckboxField);

var _CheckboxGroupField = require('./CheckboxGroupField');

var _CheckboxGroupField2 = _interopRequireDefault(_CheckboxGroupField);

var _ColorPickerField = require('./ColorPickerField');

var _ColorPickerField2 = _interopRequireDefault(_ColorPickerField);

var _DateRangePickerField = require('./DateRangePickerField');

var _DateRangePickerField2 = _interopRequireDefault(_DateRangePickerField);

var _InputField = require('./InputField');

var _InputField2 = _interopRequireDefault(_InputField);

var _NumberInputField = require('./NumberInputField');

var _NumberInputField2 = _interopRequireDefault(_NumberInputField);

var _RadioGroupField = require('./RadioGroupField');

var _RadioGroupField2 = _interopRequireDefault(_RadioGroupField);

var _SelectField = require('./SelectField');

var _SelectField2 = _interopRequireDefault(_SelectField);

var _SwitchField = require('./SwitchField');

var _SwitchField2 = _interopRequireDefault(_SwitchField);

var _Field = require('../Field');

var _Field2 = _interopRequireDefault(_Field);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var FormCheckboxField = exports.FormCheckboxField = function (_ref) {
  (0, _inherits3['default'])(FormCheckboxField, _ref);

  function FormCheckboxField() {
    (0, _classCallCheck3['default'])(this, FormCheckboxField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormCheckboxField.__proto__ || Object.getPrototypeOf(FormCheckboxField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormCheckboxField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _CheckboxField2['default'] }));
    }
  }]);
  return FormCheckboxField;
}(_react.PureComponent || _react.Component);

var FormCheckboxGroupField = exports.FormCheckboxGroupField = function (_ref2) {
  (0, _inherits3['default'])(FormCheckboxGroupField, _ref2);

  function FormCheckboxGroupField() {
    (0, _classCallCheck3['default'])(this, FormCheckboxGroupField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormCheckboxGroupField.__proto__ || Object.getPrototypeOf(FormCheckboxGroupField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormCheckboxGroupField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _CheckboxGroupField2['default'] }));
    }
  }]);
  return FormCheckboxGroupField;
}(_react.PureComponent || _react.Component);

var FormColorPickerField = exports.FormColorPickerField = function (_ref3) {
  (0, _inherits3['default'])(FormColorPickerField, _ref3);

  function FormColorPickerField() {
    (0, _classCallCheck3['default'])(this, FormColorPickerField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormColorPickerField.__proto__ || Object.getPrototypeOf(FormColorPickerField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormColorPickerField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _ColorPickerField2['default'] }));
    }
  }]);
  return FormColorPickerField;
}(_react.PureComponent || _react.Component);

var FormDateRangePickerField = exports.FormDateRangePickerField = function (_ref4) {
  (0, _inherits3['default'])(FormDateRangePickerField, _ref4);

  function FormDateRangePickerField() {
    (0, _classCallCheck3['default'])(this, FormDateRangePickerField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormDateRangePickerField.__proto__ || Object.getPrototypeOf(FormDateRangePickerField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormDateRangePickerField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _DateRangePickerField2['default'] }));
    }
  }]);
  return FormDateRangePickerField;
}(_react.PureComponent || _react.Component);

var FormInputField = exports.FormInputField = function (_ref5) {
  (0, _inherits3['default'])(FormInputField, _ref5);

  function FormInputField() {
    (0, _classCallCheck3['default'])(this, FormInputField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormInputField.__proto__ || Object.getPrototypeOf(FormInputField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormInputField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _InputField2['default'] }));
    }
  }]);
  return FormInputField;
}(_react.PureComponent || _react.Component);

var FormNumberInputField = exports.FormNumberInputField = function (_ref6) {
  (0, _inherits3['default'])(FormNumberInputField, _ref6);

  function FormNumberInputField() {
    (0, _classCallCheck3['default'])(this, FormNumberInputField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormNumberInputField.__proto__ || Object.getPrototypeOf(FormNumberInputField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormNumberInputField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _NumberInputField2['default'] }));
    }
  }]);
  return FormNumberInputField;
}(_react.PureComponent || _react.Component);

var FormRadioGroupField = exports.FormRadioGroupField = function (_ref7) {
  (0, _inherits3['default'])(FormRadioGroupField, _ref7);

  function FormRadioGroupField() {
    (0, _classCallCheck3['default'])(this, FormRadioGroupField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormRadioGroupField.__proto__ || Object.getPrototypeOf(FormRadioGroupField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormRadioGroupField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _RadioGroupField2['default'] }));
    }
  }]);
  return FormRadioGroupField;
}(_react.PureComponent || _react.Component);

var FormSelectField = exports.FormSelectField = function (_ref8) {
  (0, _inherits3['default'])(FormSelectField, _ref8);

  function FormSelectField() {
    (0, _classCallCheck3['default'])(this, FormSelectField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormSelectField.__proto__ || Object.getPrototypeOf(FormSelectField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormSelectField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _SelectField2['default'] }));
    }
  }]);
  return FormSelectField;
}(_react.PureComponent || _react.Component);

var FormSwitchField = exports.FormSwitchField = function (_ref9) {
  (0, _inherits3['default'])(FormSwitchField, _ref9);

  function FormSwitchField() {
    (0, _classCallCheck3['default'])(this, FormSwitchField);
    return (0, _possibleConstructorReturn3['default'])(this, (FormSwitchField.__proto__ || Object.getPrototypeOf(FormSwitchField)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormSwitchField, [{
    key: 'render',
    value: function render() {
      return _react2['default'].createElement(_Field2['default'], (0, _extends3['default'])({}, this.props, { component: _SwitchField2['default'] }));
    }
  }]);
  return FormSwitchField;
}(_react.PureComponent || _react.Component);
