// zent/lib/select/trigger/InputTrigger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _i18n = require('../../i18n');

var _default = require('../../i18n/default');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var InputTrigger = function (_Component) {
  (0, _inherits3['default'])(InputTrigger, _Component);

  function InputTrigger() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, InputTrigger);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref = InputTrigger.__proto__ || Object.getPrototypeOf(InputTrigger)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      value: ''
    }, _this.inputChangeHandler = function (ev) {
      _this.props.onChange({
        keyword: ev.target.value
      });
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(InputTrigger, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.props.onChange({
        extraFilter: true
      });
    }
  }, {
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps, nextState) {
      return nextState.value !== this.state.value;
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.setState({
        value: nextProps.keyword === null ? nextProps.value : nextProps.keyword
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          prefixCls = _props.prefixCls,
          placeholder = _props.placeholder,
          keyword = _props.keyword,
          text = _props.text;


      return _react2['default'].createElement(
        _i18n.I18nReceiver,
        { componentName: 'Select', defaultI18n: _default.Select },
        function (i18n) {
          return _react2['default'].createElement('input', {
            ref: function ref(input) {
              return _this2.input = input;
            },
            className: prefixCls + '-input',
            placeholder: placeholder || i18n.input,
            type: 'text',
            value: keyword === null ? text : keyword,
            onChange: _this2.inputChangeHandler,
            onClick: _this2.props.onClick
          });
        }
      );
    }
  }]);
  return InputTrigger;
}(_react.Component);

InputTrigger.propTypes = {
  prefixCls: _propTypes2['default'].string,
  value: _propTypes2['default'].any,
  placeholder: _propTypes2['default'].string
};

exports['default'] = InputTrigger;
