// zent/lib/select/trigger/TagsTrigger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

var _i18n = require('../../i18n');

var _default = require('../../i18n/default');

var _Tag = require('../components/Tag');

var _Tag2 = _interopRequireDefault(_Tag);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var TagsTrigger = function (_ref) {
  (0, _inherits3['default'])(TagsTrigger, _ref);

  function TagsTrigger(props) {
    (0, _classCallCheck3['default'])(this, TagsTrigger);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (TagsTrigger.__proto__ || Object.getPrototypeOf(TagsTrigger)).call(this, props));

    _this.deleteTagHandler = _this.deleteTagHandler.bind(_this);
    return _this;
  }

  (0, _createClass3['default'])(TagsTrigger, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var selectedItems = this.props.selectedItems;
      var cid = nextProps.cid,
          text = nextProps.text,
          value = nextProps.value;


      if (this.isDelete || this.isAdded) {
        this.isDelete = false;
        this.isAdded = false;
        return;
      }

      var isExist = selectedItems.filter(function (item) {
        return item.cid === cid;
      }).length > 0;

      if (typeof cid !== 'undefined' && !isExist) {
        selectedItems.push({
          cid: cid,
          text: text,
          value: value
        });
        this.props.onChange({
          selectedItems: selectedItems,
          selectedItem: {
            value: ''
          },
          open: false
        });
      } else if (isExist) {
        this.isAdded = true;
        this.props.onChange({
          selectedItem: {
            value: ''
          }
        });
      }
    }
  }, {
    key: 'deleteTagHandler',
    value: function deleteTagHandler(cid) {
      var selectedItems = this.props.selectedItems;

      var deleteItem = selectedItems.filter(function (item) {
        return item.cid === cid;
      })[0];
      this.isDelete = true;
      this.props.onDelete(deleteItem);
      this.props.onChange({
        selectedItems: selectedItems.filter(function (item) {
          return item.cid !== cid;
        }),
        selectedItem: {},
        open: false
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          prefixCls = _props.prefixCls,
          placeholder = _props.placeholder,
          onClick = _props.onClick,
          selectedItems = _props.selectedItems;


      return _react2['default'].createElement(
        _i18n.I18nReceiver,
        { componentName: 'Select', defaultI18n: _default.Select },
        function (i18n) {
          return _react2['default'].createElement(
            'div',
            { className: prefixCls + '-tags', onClick: onClick },
            selectedItems.length > 0 ? selectedItems.map(function (item, index) {
              return _react2['default'].createElement(_Tag2['default'], (0, _extends3['default'])({}, _this2.props, {
                key: index,
                cid: item.cid
              }, item, {
                onDelete: _this2.deleteTagHandler
              }));
            }) : placeholder || i18n.input
          );
        }
      );
    }
  }]);
  return TagsTrigger;
}(_react.PureComponent || _react.Component);

TagsTrigger.propTypes = {
  prefixCls: _propTypes2['default'].string,
  selectedItems: _propTypes2['default'].array,
  selectedItem: _propTypes2['default'].object,
  placeholder: _propTypes2['default'].string,
  value: _propTypes2['default'].any,
  onDelete: _propTypes2['default'].func
};

TagsTrigger.defaultProps = {
  selectedItems: [],
  onDelete: _noop2['default']
};

exports['default'] = TagsTrigger;
