// zent/lib/design/editor/DesignEditorItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var DesignEditorItem = function (_ref) {
  (0, _inherits3['default'])(DesignEditorItem, _ref);

  function DesignEditorItem() {
    (0, _classCallCheck3['default'])(this, DesignEditorItem);
    return (0, _possibleConstructorReturn3['default'])(this, (DesignEditorItem.__proto__ || Object.getPrototypeOf(DesignEditorItem)).apply(this, arguments));
  }

  (0, _createClass3['default'])(DesignEditorItem, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          disabled = _props.disabled,
          prefix = _props.prefix,
          className = _props.className;


      return _react2['default'].createElement(
        'div',
        { className: (0, _classnames2['default'])(prefix + '-design-editor-item', className) },
        disabled && _react2['default'].createElement('div', { className: prefix + '-design__disabled-mask' }),
        this.props.children
      );
    }
  }, {
    key: 'getBoundingBox',
    value: function getBoundingBox() {
      var node = (0, _reactDom.findDOMNode)(this);
      return node && node.getBoundingClientRect();
    }
  }]);
  return DesignEditorItem;
}(_react.PureComponent || _react.Component);

DesignEditorItem.propTypes = {
  children: _propTypes2['default'].node.isRequired,

  disabled: _propTypes2['default'].bool,

  prefix: _propTypes2['default'].string,

  className: _propTypes2['default'].string
};
DesignEditorItem.defaultProps = {
  disabled: false,
  prefix: 'zent'
};
exports['default'] = DesignEditorItem;
