// zent/lib/input/Textarea.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _omit = require('lodash/omit');

var _omit2 = _interopRequireDefault(_omit);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _autosize = require('autosize');

var _autosize2 = _interopRequireDefault(_autosize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Textarea = function (_ref) {
  (0, _inherits3['default'])(Textarea, _ref);

  function Textarea() {
    (0, _classCallCheck3['default'])(this, Textarea);
    return (0, _possibleConstructorReturn3['default'])(this, (Textarea.__proto__ || Object.getPrototypeOf(Textarea)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Textarea, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var autoSize = this.props.inputProps.autoSize;

      autoSize && (0, _autosize2['default'])(this.textarea);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      var autoSize = this.props.inputProps.autoSize;

      autoSize && _autosize2['default'].destroy(this.textarea);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          wrapClass = _props.wrapClass,
          widthStyle = _props.widthStyle,
          prefix = _props.prefix,
          handleKeyDown = _props.handleKeyDown,
          inputRef = _props.inputRef;
      var inputProps = this.props.inputProps;
      var _inputProps = inputProps,
          showCount = _inputProps.showCount,
          _inputProps$value = _inputProps.value,
          value = _inputProps$value === undefined ? '' : _inputProps$value,
          maxLength = _inputProps.maxLength;

      inputProps = (0, _omit2['default'])(inputProps, ['type', 'showCount', 'autoSize']);
      var currentCount = value.length;
      currentCount = currentCount > maxLength ? maxLength : currentCount;

      return _react2['default'].createElement(
        'div',
        { className: wrapClass, style: widthStyle },
        _react2['default'].createElement('textarea', (0, _extends3['default'])({
          ref: function ref(_ref2) {
            inputRef.input = _ref2;
            _this2.textarea = _ref2;
          },
          className: (0, _classnames2['default'])(prefix + '-textarea', (0, _defineProperty3['default'])({}, prefix + '-textarea-with-count', showCount))
        }, inputProps, {
          onKeyDown: handleKeyDown
        })),
        showCount && _react2['default'].createElement(
          'span',
          { className: prefix + '-textarea-count' },
          currentCount,
          '/',
          maxLength
        )
      );
    }
  }]);
  return Textarea;
}(_react.PureComponent || _react.Component);

exports['default'] = Textarea;
