// zent/lib/card/Card.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _isNil = require('lodash/isNil');

var _isNil2 = _interopRequireDefault(_isNil);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Card = function (_ref) {
  (0, _inherits3['default'])(Card, _ref);

  function Card() {
    (0, _classCallCheck3['default'])(this, Card);
    return (0, _possibleConstructorReturn3['default'])(this, (Card.__proto__ || Object.getPrototypeOf(Card)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Card, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          title = _props.title,
          action = _props.action,
          style = _props.style,
          children = _props.children,
          className = _props.className,
          bodyStyle = _props.bodyStyle,
          prefix = _props.prefix;


      this.isValidTitle = !(0, _isNil2['default'])(title);
      this.isValidAction = !(0, _isNil2['default'])(action);

      return _react2['default'].createElement(
        'div',
        { className: prefix + '-card ' + className, style: style },
        (this.isValidTitle || this.isValidAction) && _react2['default'].createElement(
          'div',
          { className: prefix + '-card-header' },
          this.isValidTitle && _react2['default'].createElement(
            'h3',
            { className: prefix + '-card-header__title' },
            title
          ),
          this.isValidAction && _react2['default'].createElement(
            'div',
            { className: prefix + '-card-header__action' },
            action
          )
        ),
        _react2['default'].createElement(
          'div',
          { className: prefix + '-card-body', style: bodyStyle },
          children
        )
      );
    }
  }]);
  return Card;
}(_react.PureComponent || _react.Component);

Card.defaultProps = {
  style: {},
  bodyStyle: {},
  className: '',
  prefix: 'zent'
};
Card.propTypes = {
  title: _propTypes2['default'].node,
  action: _propTypes2['default'].node,
  style: _propTypes2['default'].object,
  bodyStyle: _propTypes2['default'].object,
  className: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string
};
exports['default'] = Card;
