// zent/lib/datetimepicker/quarter/QuarterPanelBody.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _PanelCell = require('../common/PanelCell');

var _PanelCell2 = _interopRequireDefault(_PanelCell);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var ROW = 2;
var COL = 2;

var QuarterPanelBody = function (_ref) {
  (0, _inherits3['default'])(QuarterPanelBody, _ref);

  function QuarterPanelBody() {
    (0, _classCallCheck3['default'])(this, QuarterPanelBody);
    return (0, _possibleConstructorReturn3['default'])(this, (QuarterPanelBody.__proto__ || Object.getPrototypeOf(QuarterPanelBody)).apply(this, arguments));
  }

  (0, _createClass3['default'])(QuarterPanelBody, [{
    key: 'getQuarters',
    value: function getQuarters() {
      var _props = this.props,
          disabledDate = _props.disabledDate,
          selected = _props.selected,
          i18n = _props.i18n;

      var quarters = [];
      var index = 0;
      for (var rowIndex = 0; rowIndex < ROW; rowIndex++) {
        quarters[rowIndex] = [];
        for (var colIndex = 0; colIndex < COL; colIndex++) {
          var isSelected = selected && (0, _utils.getQuarterFromDate)(selected) === index;
          var isDisabled = disabledDate && disabledDate(index);
          var className = (0, _classnames2['default'])({
            'panel__cell quarter-panel__cell': true,
            'panel__cell--current': false,
            'panel__cell--selected': isSelected,
            'panel__cell--disabled': isDisabled
          });
          quarters[rowIndex][colIndex] = {
            text: i18n.panel.quarterNames[index],
            value: index,
            title: i18n.panel.quarterNames[index],
            className: className,
            isDisabled: isDisabled
          };
          index++;
        }
      }

      return quarters;
    }
  }, {
    key: 'render',
    value: function render() {
      var onSelect = this.props.onSelect;

      var quarters = this.getQuarters();

      return _react2['default'].createElement(
        'div',
        { className: 'quarter-table panel-table' },
        _react2['default'].createElement(_PanelCell2['default'], { onSelect: onSelect, cells: quarters })
      );
    }
  }]);
  return QuarterPanelBody;
}(_react.PureComponent || _react.Component);

exports['default'] = QuarterPanelBody;
