// zent/lib/datetimepicker/quarter/QuarterPanel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _PanelHeader = require('../common/PanelHeader');

var _PanelHeader2 = _interopRequireDefault(_PanelHeader);

var _QuarterPanelBody = require('./QuarterPanelBody');

var _QuarterPanelBody2 = _interopRequireDefault(_QuarterPanelBody);

var _YearPanel = require('../year/YearPanel');

var _YearPanel2 = _interopRequireDefault(_YearPanel);

var _utils = require('../utils/');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var QuarterPanel = function (_ref) {
  (0, _inherits3['default'])(QuarterPanel, _ref);

  function QuarterPanel() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, QuarterPanel);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = QuarterPanel.__proto__ || Object.getPrototypeOf(QuarterPanel)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      showYear: false
    }, _this.prevYear = function () {
      var _this$props = _this.props,
          actived = _this$props.actived,
          onChange = _this$props.onChange;

      var prev = (0, _utils.goYears)(actived, -1);
      onChange(prev);
    }, _this.nextYear = function () {
      var _this$props2 = _this.props,
          actived = _this$props2.actived,
          onChange = _this$props2.onChange;

      var next = (0, _utils.goYears)(actived, 1);
      onChange(next);
    }, _this.showYearPanel = function () {
      _this.setState({
        showYear: true
      });
    }, _this.onSelectYear = function (val) {
      var close = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var _this$props3 = _this.props,
          actived = _this$props3.actived,
          onChange = _this$props3.onChange;

      var acp = new Date(actived);

      acp.setFullYear(val);
      onChange(acp);

      _this.setState({
        showYear: close
      });
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(QuarterPanel, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          actived = _props.actived,
          disabledDate = _props.disabledDate,
          i18n = _props.i18n,
          onSelect = _props.onSelect,
          selected = _props.selected,
          showYear = this.state.showYear;


      var title = '' + actived.getFullYear();

      var yearPanel = void 0;
      if (showYear) {
        yearPanel = _react2['default'].createElement(_YearPanel2['default'], {
          actived: actived,
          selected: selected,
          onChange: this.onSelectYear,
          onSelect: this.onSelectYear
        });
      }

      return _react2['default'].createElement(
        'div',
        { className: 'quarter-panel' },
        _react2['default'].createElement(_PanelHeader2['default'], {
          title: title,
          onClickTitle: this.showYearPanel,
          prev: this.prevYear,
          next: this.nextYear
        }),
        _react2['default'].createElement(_QuarterPanelBody2['default'], {
          actived: actived,
          selected: selected,
          disabledDate: disabledDate,
          onSelect: onSelect,
          i18n: i18n
        }),
        showYear && yearPanel
      );
    }
  }]);
  return QuarterPanel;
}(_react.PureComponent || _react.Component);

exports['default'] = QuarterPanel;
