// zent/lib/datetimepicker/SplitDateRangePicker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _isString = require('lodash/isString');

var _isString2 = _interopRequireDefault(_isString);

var _isDate = require('lodash/isDate');

var _isDate2 = _interopRequireDefault(_isDate);

var _i18n = require('../i18n');

var _default = require('../i18n/default');

var _constants = require('./constants');

var _DatePicker = require('./DatePicker');

var _DatePicker2 = _interopRequireDefault(_DatePicker);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

// type
var START = 'start';
var END = 'end';

var SplitDateRangePicker = function (_ref) {
  (0, _inherits3['default'])(SplitDateRangePicker, _ref);

  function SplitDateRangePicker() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, SplitDateRangePicker);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = SplitDateRangePicker.__proto__ || Object.getPrototypeOf(SplitDateRangePicker)).call.apply(_ref2, [this].concat(args))), _this), _this.onChange = function (type) {
      return function (val) {
        var _this$props = _this.props,
            onChange = _this$props.onChange,
            value = _this$props.value;

        var ret = value ? value.slice() : [];

        if (type === START) {
          ret.length === 2 ? ret.splice(0, 1, val) : ret.splice(0, 1, val, '');
        } else {
          ret.length === 0 ? ret.splice(1, 1, '', val) : ret.splice(1, 1, val);
        }
        onChange(ret);
      };
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(SplitDateRangePicker, [{
    key: 'renderPicker',
    value: function renderPicker() {
      var _this2 = this;

      var _props = this.props,
          className = _props.className,
          defaultTime = _props.defaultTime,
          _disabledDate = _props.disabledDate,
          onChange = _props.onChange,
          _onClick = _props.onClick,
          _onClose = _props.onClose,
          _onOpen = _props.onOpen,
          openPanel = _props.openPanel,
          placeholder = _props.placeholder,
          value = _props.value,
          pickerProps = (0, _objectWithoutProperties3['default'])(_props, ['className', 'defaultTime', 'disabledDate', 'onChange', 'onClick', 'onClose', 'onOpen', 'openPanel', 'placeholder', 'value']);

      var rangePicker = void 0;
      // 兼容老 api ，支持传入字符串
      var timeArr = (0, _isString2['default'])(defaultTime) || (0, _isDate2['default'])(defaultTime) ? [defaultTime, defaultTime] : defaultTime;
      var pickerCls = (0, _classnames2['default'])('range-picker2');

      rangePicker = _react2['default'].createElement(
        'div',
        { className: pickerCls },
        _react2['default'].createElement(
          _i18n.I18nReceiver,
          { componentName: 'TimePicker', defaultI18n: _default.TimePicker },
          function (i18n) {
            return _react2['default'].createElement(_DatePicker2['default'], (0, _extends3['default'])({}, pickerProps, {
              openPanel: openPanel[0],
              placeholder: placeholder[0] || i18n.start,
              max: value[1] || pickerProps.max,
              defaultTime: timeArr[0],
              value: value[0],
              onClick: function onClick(val) {
                return _onClick && _onClick(val, START);
              },
              onOpen: function onOpen() {
                return _onOpen && _onOpen(START);
              },
              onClose: function onClose() {
                return _onClose && _onClose(START);
              },
              onChange: _this2.onChange(START),
              disabledDate: function disabledDate(val) {
                return _disabledDate(val, START);
              }
            }));
          }
        ),
        _react2['default'].createElement(
          _i18n.I18nReceiver,
          { componentName: 'TimePicker', defaultI18n: _default.TimePicker },
          function (i18n) {
            return _react2['default'].createElement(
              'span',
              { className: 'picker-seperator' },
              i18n.to
            );
          }
        ),
        _react2['default'].createElement(
          _i18n.I18nReceiver,
          { componentName: 'TimePicker', defaultI18n: _default.TimePicker },
          function (i18n) {
            return _react2['default'].createElement(_DatePicker2['default'], (0, _extends3['default'])({}, pickerProps, {
              openPanel: openPanel[1],
              placeholder: placeholder[1] || i18n.end,
              min: value[0] || pickerProps.min,
              defaultTime: timeArr[1],
              value: value[1],
              onClick: function onClick(val) {
                return _onClick && _onClick(val, END);
              },
              onOpen: function onOpen() {
                return _onOpen && _onOpen(END);
              },
              onClose: function onClose() {
                return _onClose && _onClose(END);
              },
              onChange: _this2.onChange(END),
              disabledDate: function disabledDate(val) {
                return _disabledDate(val, END);
              }
            }));
          }
        )
      );

      return rangePicker;
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          prefix = _props2.prefix,
          className = _props2.className;

      var prefixCls = prefix + '-datetime-picker ' + className;

      return _react2['default'].createElement(
        'div',
        { className: prefixCls },
        this.renderPicker()
      );
    }
  }]);
  return SplitDateRangePicker;
}(_react.PureComponent || _react.Component);

SplitDateRangePicker.propTypes = (0, _extends3['default'])({}, _constants.commonPropTypes, {
  showTime: _propTypes2['default'].bool,
  placeholder: _propTypes2['default'].array,
  defaultTime: _propTypes2['default'].arrayOf(_propTypes2['default'].oneOfType([_propTypes2['default'].string, _propTypes2['default'].instanceOf(Date)])),
  valueType: _propTypes2['default'].oneOf(['string', 'number', 'date'])
});
SplitDateRangePicker.defaultProps = (0, _extends3['default'])({}, _constants.commonProps, {
  placeholder: ['', ''],
  format: 'YYYY-MM-DD',
  value: [],
  openPanel: [],
  defaultTime: [_utils.TIME_BEGIN, _utils.TIME_BEGIN]
});
exports['default'] = SplitDateRangePicker;
