// zent/lib/grid/Body.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _get = require('lodash/get');

var _get2 = _interopRequireDefault(_get);

var _forEach = require('lodash/forEach');

var _forEach2 = _interopRequireDefault(_forEach);

var _Row = require('./Row');

var _Row2 = _interopRequireDefault(_Row);

var _ColGroup = require('./ColGroup');

var _ColGroup2 = _interopRequireDefault(_ColGroup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Body = function (_ref) {
  (0, _inherits3['default'])(Body, _ref);

  function Body() {
    (0, _classCallCheck3['default'])(this, Body);
    return (0, _possibleConstructorReturn3['default'])(this, (Body.__proto__ || Object.getPrototypeOf(Body)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Body, [{
    key: 'getRows',
    value: function getRows() {
      var _props = this.props,
          prefix = _props.prefix,
          datasets = _props.datasets,
          columns = _props.columns,
          rowKey = _props.rowKey,
          rowClassName = _props.rowClassName,
          onRowClick = _props.onRowClick,
          fixed = _props.fixed,
          fixedColumnsBodyRowsHeight = _props.fixedColumnsBodyRowsHeight;

      var row = [];

      (0, _forEach2['default'])(datasets, function (data, index) {
        row.push(_react2['default'].createElement(_Row2['default'], {
          data: data,
          columns: columns,
          index: index,
          rowIndex: index,
          prefix: prefix,
          key: rowKey ? (0, _get2['default'])(data, rowKey) : index,
          rowClassName: rowClassName,
          onRowClick: onRowClick,
          fixed: fixed,
          fixedColumnsBodyRowsHeight: fixedColumnsBodyRowsHeight
        }));
      });

      return row;
    }
  }, {
    key: 'renderTbody',
    value: function renderTbody() {
      var prefix = this.props.prefix;


      return _react2['default'].createElement(
        'tbody',
        { className: prefix + '-grid-tbody' },
        this.getRows()
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          scroll = _props2.scroll,
          fixed = _props2.fixed,
          prefix = _props2.prefix,
          columns = _props2.columns;

      var bodyStyle = {};
      if (!fixed && scroll.x) {
        bodyStyle.width = scroll.x;
      }
      return scroll.y ? _react2['default'].createElement(
        'table',
        { className: prefix + '-grid-table', style: bodyStyle },
        _react2['default'].createElement(_ColGroup2['default'], { columns: columns }),
        this.renderTbody()
      ) : this.renderTbody();
    }
  }]);
  return Body;
}(_react.PureComponent || _react.Component);

Body.propTypes = {
  prefix: _propTypes2['default'].string,
  columns: _propTypes2['default'].array
};

exports['default'] = Body;
