// zent/lib/grid/Row.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _forEach = require('lodash/forEach');

var _forEach2 = _interopRequireDefault(_forEach);

var _isFunction = require('lodash/isFunction');

var _isFunction2 = _interopRequireDefault(_isFunction);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Cell = require('./Cell');

var _Cell2 = _interopRequireDefault(_Cell);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Row = function (_ref) {
  (0, _inherits3['default'])(Row, _ref);

  function Row() {
    (0, _classCallCheck3['default'])(this, Row);
    return (0, _possibleConstructorReturn3['default'])(this, (Row.__proto__ || Object.getPrototypeOf(Row)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Row, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          prefix = _props.prefix,
          columns = _props.columns,
          data = _props.data,
          rowIndex = _props.rowIndex,
          rowClassName = _props.rowClassName,
          onRowClick = _props.onRowClick,
          fixed = _props.fixed,
          fixedColumnsBodyRowsHeight = _props.fixedColumnsBodyRowsHeight;


      var cells = [];

      var className = (0, _isFunction2['default'])(rowClassName) ? rowClassName(data, rowIndex) : rowClassName;

      var height = fixed && fixedColumnsBodyRowsHeight[rowIndex] ? fixedColumnsBodyRowsHeight[rowIndex] : null;

      (0, _forEach2['default'])(columns, function (column, columnIndex) {
        var pos = {
          row: rowIndex,
          column: columnIndex
        };

        cells.push(_react2['default'].createElement(_Cell2['default'], {
          column: column,
          data: data,
          pos: pos,
          columnIndex: columnIndex,
          key: columnIndex,
          prefix: prefix
        }));
      });

      return _react2['default'].createElement(
        'tr',
        {
          className: (0, _classnames2['default'])(prefix + '-grid-tr', className),
          onClick: function onClick(e) {
            return onRowClick(data, rowIndex, e);
          },
          style: { height: height }
        },
        cells
      );
    }
  }]);
  return Row;
}(_react.PureComponent || _react.Component);

exports['default'] = Row;
