// zent/lib/grid/SelectionCheckbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _checkbox = require('../checkbox');

var _checkbox2 = _interopRequireDefault(_checkbox);

var _includes = require('lodash/includes');

var _includes2 = _interopRequireDefault(_includes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var SelectionCheckbox = function (_ref) {
  (0, _inherits3['default'])(SelectionCheckbox, _ref);

  function SelectionCheckbox(props) {
    (0, _classCallCheck3['default'])(this, SelectionCheckbox);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (SelectionCheckbox.__proto__ || Object.getPrototypeOf(SelectionCheckbox)).call(this, props));

    _initialiseProps.call(_this);

    _this.state = {
      checked: _this.getCheckState(props)
    };
    return _this;
  }

  (0, _createClass3['default'])(SelectionCheckbox, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.subscribe();
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps, nextState) {
      var checked = this.getCheckState(nextProps);
      if (checked !== nextState.checked) {
        this.setState({ checked: checked });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.unsubscribe) {
        this.unsubscribe();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          onChange = _props.onChange,
          disabled = _props.disabled;
      var checked = this.state.checked;

      return _react2['default'].createElement(_checkbox2['default'], { onChange: onChange, checked: checked, disabled: disabled });
    }
  }]);
  return SelectionCheckbox;
}(_react.PureComponent || _react.Component);

var _initialiseProps = function _initialiseProps() {
  var _this2 = this;

  this.subscribe = function () {
    var store = _this2.props.store;

    _this2.unsubscribe = store.subscribe('selectedRowKeys', function () {
      var checked = _this2.getCheckState(_this2.props);
      if (_this2.state.checked !== checked) {
        _this2.setState({ checked: checked });
      }
    });
  };

  this.getCheckState = function (props) {
    var store = props.store,
        rowIndex = props.rowIndex;

    return (0, _includes2['default'])(store.getState('selectedRowKeys'), rowIndex);
  };
};

exports['default'] = SelectionCheckbox;
