// zent/lib/upload/utils/sortable.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

exports.initSortable = initSortable;
exports.swapArray = swapArray;

var _sortablejs = require('sortablejs');

var _sortablejs2 = _interopRequireDefault(_sortablejs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var UNSORTABLE = 'unsortable';

function initSortable(el, onMove) {
  var initState = [];
  var sortable = _sortablejs2['default'].create(el, {
    filter: '.' + UNSORTABLE,
    onStart: function onStart() {
      initState = sortable.toArray();
    },
    onMove: function onMove(evt) {
      return evt.related.className !== UNSORTABLE;
    },
    onEnd: function onEnd(evt) {
      var newIndex = evt.newIndex,
          oldIndex = evt.oldIndex;

      sortable.sort(initState);
      onMove(oldIndex, newIndex);
    }
  });
  return sortable;
}

function swapArray(list, fromIndex, toIndex) {
  var result = Array.from([].concat(list));

  var _result$splice = result.splice(fromIndex, 1),
      _result$splice2 = (0, _slicedToArray3['default'])(_result$splice, 1),
      removed = _result$splice2[0];

  result.splice(toIndex, 0, removed);
  return result;
}
