// zent/lib/datetimepicker/month/MonthPanelBody.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _PanelCell = require('../common/PanelCell');

var _PanelCell2 = _interopRequireDefault(_PanelCell);

var _utils = require('../utils/');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var ROW = 4;
var COL = 3;

var MonthPanelBody = function (_ref) {
  (0, _inherits3['default'])(MonthPanelBody, _ref);

  function MonthPanelBody() {
    (0, _classCallCheck3['default'])(this, MonthPanelBody);
    return (0, _possibleConstructorReturn3['default'])(this, (MonthPanelBody.__proto__ || Object.getPrototypeOf(MonthPanelBody)).apply(this, arguments));
  }

  (0, _createClass3['default'])(MonthPanelBody, [{
    key: 'isSelected',
    value: function isSelected(val) {
      var _props = this.props,
          selected = _props.selected,
          year = _props.year;


      if (selected && selected instanceof Date) {
        return val === selected.getMonth() && year == selected.getFullYear();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props2 = this.props,
          onSelect = _props2.onSelect,
          disabledDate = _props2.disabledDate,
          i18n = _props2.i18n,
          year = _props2.year;


      return _react2['default'].createElement(
        'div',
        { className: 'month-table panel-table' },
        _react2['default'].createElement(_PanelCell2['default'], {
          onSelect: onSelect,
          cells: function () {
            var months = [],
                curYear = new Date().getFullYear();
            var index = 0;
            for (var rowIndex = 0; rowIndex < ROW; rowIndex++) {
              months[rowIndex] = [];
              for (var colIndex = 0; colIndex < COL; colIndex++) {
                var isCurrent = index === _utils.CURRENT_MONTH && year == curYear;
                var isSelected = _this2.isSelected(index);
                var isDisabled = disabledDate && disabledDate(index);
                var className = (0, _classnames2['default'])({
                  'panel__cell month-panel__cell': true,
                  'panel__cell--current': isCurrent,
                  'panel__cell--selected': isSelected,
                  'panel__cell--disabled': isDisabled
                });
                months[rowIndex][colIndex] = {
                  text: i18n.panel.monthNames[index],
                  value: index,
                  title: i18n.panel.monthNames[index],
                  className: className,
                  isDisabled: isDisabled
                };
                index++;
              }
            }
            return months;
          }()
        })
      );
    }
  }]);
  return MonthPanelBody;
}(_react.PureComponent || _react.Component);

exports['default'] = MonthPanelBody;
