// zent/lib/form/FormSection.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var FormSection = function (_ref) {
  (0, _inherits3['default'])(FormSection, _ref);

  function FormSection(props, context) {
    (0, _classCallCheck3['default'])(this, FormSection);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (FormSection.__proto__ || Object.getPrototypeOf(FormSection)).call(this, props, context));

    if (!context.zentForm) {
      throw new Error('FormSection must be in zent-form');
    }
    return _this;
  }

  (0, _createClass3['default'])(FormSection, [{
    key: 'getChildContext',
    value: function getChildContext() {
      var zentForm = this.context.zentForm;
      var name = this.props.name;

      return {
        zentForm: (0, _extends3['default'])({}, zentForm, {
          prefix: (0, _utils.prefixName)(zentForm, name)
        })
      };
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          component = _props.component,
          rest = (0, _objectWithoutProperties3['default'])(_props, ['children', 'component']);

      return (0, _react.createElement)(component, (0, _extends3['default'])({}, rest, {
        children: children
      }));
    }
  }]);
  return FormSection;
}(_react.PureComponent || _react.Component);

FormSection.propTypes = {
  name: _propTypes2['default'].string.isRequired,
  component: _propTypes2['default'].oneOfType([_propTypes2['default'].element, _propTypes2['default'].string]),
  children: _propTypes2['default'].any
};
FormSection.defaultProps = {
  component: 'div'
};
FormSection.contextTypes = {
  zentForm: _propTypes2['default'].object
};
FormSection.childContextTypes = {
  zentForm: _propTypes2['default'].object.isRequired
};
exports['default'] = FormSection;
