// zent/lib/form/Form.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Form = function (_ref) {
  (0, _inherits3['default'])(Form, _ref);

  function Form() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Form);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = Form.__proto__ || Object.getPrototypeOf(Form)).call.apply(_ref2, [this].concat(args))), _this), _this.onKeyDown = function (event) {
      // 默认禁止回车触发表单提交事件
      if (_this.props.disableEnterSubmit && event.keyCode === 13) {
        event.preventDefault();
      }
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Form, [{
    key: 'render',
    value: function render() {
      var _classNames;

      var _props = this.props,
          prefix = _props.prefix,
          className = _props.className,
          style = _props.style,
          horizontal = _props.horizontal,
          inline = _props.inline,
          onSubmit = _props.onSubmit;

      var formClassName = (0, _classnames2['default'])((_classNames = {}, (0, _defineProperty3['default'])(_classNames, prefix + '-form', true), (0, _defineProperty3['default'])(_classNames, prefix + '-form--horizontal', horizontal), (0, _defineProperty3['default'])(_classNames, prefix + '-form--inline', inline), (0, _defineProperty3['default'])(_classNames, prefix + '-form--vertical', !horizontal && !inline), (0, _defineProperty3['default'])(_classNames, className, !!className), _classNames));
      return _react2['default'].createElement(
        'form',
        {
          className: formClassName,
          style: style,
          onSubmit: onSubmit,
          onKeyDown: this.onKeyDown
        },
        this.props.children
      );
    }
  }]);
  return Form;
}(_react.PureComponent || _react.Component);

Form.propTypes = {
  prefix: _propTypes2['default'].string,
  className: _propTypes2['default'].string,
  horizontal: _propTypes2['default'].bool,
  inline: _propTypes2['default'].bool,
  vertical: _propTypes2['default'].bool,
  onSubmit: _propTypes2['default'].func,
  children: _propTypes2['default'].any,
  style: _propTypes2['default'].object,
  disableEnterSubmit: _propTypes2['default'].bool
};
Form.defaultProps = {
  prefix: 'zent',
  onSubmit: _noop2['default'],
  disableEnterSubmit: true
};
exports['default'] = Form;
