// zent/lib/form/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('lodash/assign');

var _assign2 = _interopRequireDefault(_assign);

var _Form = require('./Form');

var _Form2 = _interopRequireDefault(_Form);

var _createForm = require('./createForm');

var _createForm2 = _interopRequireDefault(_createForm);

var _Field = require('./Field');

var _Field2 = _interopRequireDefault(_Field);

var _FieldArray = require('./FieldArray');

var _FieldArray2 = _interopRequireDefault(_FieldArray);

var _Fieldset = require('./Fieldset');

var _Fieldset2 = _interopRequireDefault(_Fieldset);

var _FormSection = require('./FormSection');

var _FormSection2 = _interopRequireDefault(_FormSection);

var _getControlGroup = require('./getControlGroup');

var _getControlGroup2 = _interopRequireDefault(_getControlGroup);

var _unknownProps = require('./unknownProps');

var _unknownProps2 = _interopRequireDefault(_unknownProps);

var _SubmissionError = require('./SubmissionError');

var _SubmissionError2 = _interopRequireDefault(_SubmissionError);

var _InputField = require('./form-components/InputField');

var _InputField2 = _interopRequireDefault(_InputField);

var _CheckboxField = require('./form-components/CheckboxField');

var _CheckboxField2 = _interopRequireDefault(_CheckboxField);

var _CheckboxGroupField = require('./form-components/CheckboxGroupField');

var _CheckboxGroupField2 = _interopRequireDefault(_CheckboxGroupField);

var _RadioGroupField = require('./form-components/RadioGroupField');

var _RadioGroupField2 = _interopRequireDefault(_RadioGroupField);

var _SelectField = require('./form-components/SelectField');

var _SelectField2 = _interopRequireDefault(_SelectField);

var _NumberInputField = require('./form-components/NumberInputField');

var _NumberInputField2 = _interopRequireDefault(_NumberInputField);

var _ColorPickerField = require('./form-components/ColorPickerField');

var _ColorPickerField2 = _interopRequireDefault(_ColorPickerField);

var _DateRangePickerField = require('./form-components/DateRangePickerField');

var _DateRangePickerField2 = _interopRequireDefault(_DateRangePickerField);

var _SwitchField = require('./form-components/SwitchField');

var _SwitchField2 = _interopRequireDefault(_SwitchField);

var _FormComponentField = require('./form-components/FormComponentField');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

exports['default'] = (0, _assign2['default'])(_Form2['default'], {
  Form: _Form2['default'],
  createForm: _createForm2['default'],
  Field: _Field2['default'],
  FieldArray: _FieldArray2['default'],
  Fieldset: _Fieldset2['default'],
  FormSection: _FormSection2['default'],
  getControlGroup: _getControlGroup2['default'],
  unknownProps: _unknownProps2['default'],
  InputField: _InputField2['default'],
  CheckboxField: _CheckboxField2['default'],
  CheckboxGroupField: _CheckboxGroupField2['default'],
  RadioGroupField: _RadioGroupField2['default'],
  SelectField: _SelectField2['default'],
  NumberInputField: _NumberInputField2['default'],
  ColorPickerField: _ColorPickerField2['default'],
  DateRangePickerField: _DateRangePickerField2['default'],
  SwitchField: _SwitchField2['default'],
  SubmissionError: _SubmissionError2['default'],
  FormCheckboxField: _FormComponentField.FormCheckboxField,
  FormCheckboxGroupField: _FormComponentField.FormCheckboxGroupField,
  FormColorPickerField: _FormComponentField.FormColorPickerField,
  FormDateRangePickerField: _FormComponentField.FormDateRangePickerField,
  FormInputField: _FormComponentField.FormInputField,
  FormNumberInputField: _FormComponentField.FormNumberInputField,
  FormRadioGroupField: _FormComponentField.FormRadioGroupField,
  FormSelectField: _FormComponentField.FormSelectField,
  FormSwitchField: _FormComponentField.FormSwitchField
});
