// zent/lib/design/components/richtext/editor/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _assign = require('lodash/assign');

var _assign2 = _interopRequireDefault(_assign);

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _uuid = require('../../../../utils/uuid');

var _uuid2 = _interopRequireDefault(_uuid);

require('./plugins');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var UEDITOR_LOADED_KEY = '__ZENT_UEDITOR_LOADED_STATUS__';

// ueditor 默认值
var initConfig = {
  toolbars: [['bold', 'italic', 'underline', 'strikethrough', 'forecolor', 'backcolor', 'justifyleft', 'justifycenter', 'justifyright', '|', 'insertunorderedlist', 'insertorderedlist', 'blockquote'], ['emotion', 'insertvideo', 'link', 'removeformat', '|', 'rowspacingtop', 'rowspacingbottom', 'lineheight', 'paragraph', 'fontsize'], ['inserttable', 'deletetable', 'insertparagraphbeforetable', 'insertrow', 'deleterow', 'insertcol', 'deletecol', 'mergecells', 'mergeright', 'mergedown', 'splittocells', 'splittorows', 'splittocols']],
  autoClearinitialContent: false,
  autoFloatEnabled: true, // 是否保持 toolbar 滚动时不动
  focus: false,
  wordCount: true,
  elementPathEnabled: false,
  pasteplain: false, // 是否默认为纯文本粘贴。false为不使用纯文本粘贴，true为使用纯文本粘贴
  initialFrameWidth: 640, // 初始化编辑器宽度
  initialFrameHeight: 200,
  maximumWords: 10000
};

var RichText = function (_Component) {
  (0, _inherits3['default'])(RichText, _Component);

  function RichText(props) {
    (0, _classCallCheck3['default'])(this, RichText);

    // 生成加载 ueditor 的节点id
    var _this = (0, _possibleConstructorReturn3['default'])(this, (RichText.__proto__ || Object.getPrototypeOf(RichText)).call(this, props));

    _this.initRichText = function () {
      var UE = window.UE;
      var target = document.getElementById(_this.uuid);

      if (!UE || !target) return false;

      var _this$props = _this.props,
          value = _this$props.value,
          editorConfig = _this$props.editorConfig;


      var conf = (0, _assign2['default'])({}, initConfig, editorConfig);

      var editor = new UE.ui.Editor(conf);
      _this.editor = editor;

      editor.addListener('blur contentChange', function () {
        _this.onChange();
      });
      editor.render(target);
      editor.ready(function () {
        editor.setContent(value);
      });
    };

    _this.onChange = function () {
      var val = _this.editor.getContent();
      _this.props.onChange && _this.props.onChange(val);
    };

    _this.uuid = (0, _uuid2['default'])();
    return _this;
  }

  (0, _createClass3['default'])(RichText, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      var timer = null;

      if (window.UE) {
        this.initRichText();
      } else {
        // 当一个页面中存在多个 RichText 组件时，需避免加载多份 ueditor.js
        timer = setInterval(function () {
          var status = window[UEDITOR_LOADED_KEY];
          if (status === 1) {
            // 加载中
          } else if (status === 2) {
            clearInterval(timer);
            _this2.initRichText();
          } else {
            _this2.loadUEditorScript();
          }
        }, 50);
      }
    }
  }, {
    key: 'loadUEditorScript',
    value: function loadUEditorScript() {
      var _this3 = this;

      if (window[UEDITOR_LOADED_KEY] !== undefined) return;

      window[UEDITOR_LOADED_KEY] = 1; // 加载中
      var _props = this.props,
          ueditorHomeUrl = _props.ueditorHomeUrl,
          ueditorIframeUrl = _props.ueditorIframeUrl,
          ueditorUrl = _props.ueditorUrl,
          ueditorConfigUrl = _props.ueditorConfigUrl;


      window.UEDITOR_HOME_URL = ueditorHomeUrl;
      window.UEDITOR_IFRAME_URL = ueditorIframeUrl;

      this.createScript(ueditorConfigUrl, function () {
        _this3.createScript(ueditorUrl, function () {
          window[UEDITOR_LOADED_KEY] = 2; // 加载完成
        });
      });
    }
  }, {
    key: 'createScript',
    value: function createScript(url, callback) {
      var oScript = document.createElement('script');
      oScript.type = 'text/javascript';
      oScript.async = true;
      oScript.src = url;

      oScript.onload = function () {
        callback();
      };

      document.body.appendChild(oScript);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (!this.editor) return;
      this.editor.destroy();
    }
  }, {
    key: 'render',
    value: function render() {
      var prefix = this.props.prefix;


      return _react2['default'].createElement(
        'div',
        { className: prefix + '-richtext ' + this.props.className },
        _react2['default'].createElement('div', { id: this.uuid })
      );
    }
  }]);
  return RichText;
}(_react.Component);

RichText.propTypes = {
  value: _propTypes2['default'].string,
  onChange: _propTypes2['default'].func,
  ueditorUrl: _propTypes2['default'].string,
  ueditorConfigUrl: _propTypes2['default'].string,
  ueditorHomeUrl: _propTypes2['default'].string,
  ueditorIframeUrl: _propTypes2['default'].string,
  editorConfig: _propTypes2['default'].object,
  className: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string
};
RichText.defaultProps = {
  value: '',
  onChange: _noop2['default'],
  ueditorUrl: '//b.yzcdn.cn/v2/vendor/ueditor/release/ueditor.all.min.201707251345.js',
  ueditorConfigUrl: '//b.yzcdn.cn/v2/vendor/ueditor/release/ueditor.config.201707251345.js',
  ueditorHomeUrl: '//b.yzcdn.cn/v2/vendor/ueditor/dist/',
  ueditorIframeUrl: '//www.youzan.com/v2/static/vendor/ueditor/dist/',
  editorConfig: {}, // ueditor 默认值
  className: '',
  prefix: 'zent-design-component'
};
exports['default'] = RichText;
