// zent/lib/radio/Group.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Group = function (_ref) {
  (0, _inherits3['default'])(Group, _ref);

  function Group() {
    (0, _classCallCheck3['default'])(this, Group);
    return (0, _possibleConstructorReturn3['default'])(this, (Group.__proto__ || Object.getPrototypeOf(Group)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Group, [{
    key: 'onRadioChange',
    value: function onRadioChange(e) {
      this.props.onChange(e);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this,
          _classNames;

      var _props = this.props,
          className = _props.className,
          prefix = _props.prefix,
          style = _props.style,
          isValueEqual = _props.isValueEqual;

      var children = _react2['default'].Children.map(this.props.children, function (radio) {
        if (radio && radio.props) {
          return _react2['default'].cloneElement(radio, (0, _extends3['default'])({}, radio.props, {
            onChange: _this2.onRadioChange.bind(_this2),
            checked: isValueEqual(_this2.props.value, radio.props.value),
            disabled: radio.props.disabled !== undefined ? radio.props.disabled : _this2.props.disabled,
            readOnly: radio.props.readOnly !== undefined ? radio.props.readOnly : _this2.props.readOnly
          }));
        }
      });

      var classString = (0, _classnames2['default'])((_classNames = {}, (0, _defineProperty3['default'])(_classNames, prefix + '-radio-group', true), (0, _defineProperty3['default'])(_classNames, className, !!className), _classNames));

      return _react2['default'].createElement(
        'div',
        { className: classString, style: style },
        children
      );
    }
  }]);
  return Group;
}(_react.PureComponent || _react.Component);

Group.propTypes = {
  value: _propTypes2['default'].any,
  isValueEqual: _propTypes2['default'].func,
  onChange: _propTypes2['default'].func,
  className: _propTypes2['default'].string,
  style: _propTypes2['default'].object,
  prefix: _propTypes2['default'].string
};
Group.defaultProps = {
  prefix: 'zent',
  className: '',
  style: {},
  isValueEqual: function isValueEqual(a, b) {
    return a === b;
  },
  onChange: function onChange() {}
};
exports['default'] = Group;
