// zent/lib/radio/Radio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _assign = require('lodash/assign');

var _assign2 = _interopRequireDefault(_assign);

var _getWidth = require('../utils/getWidth');

var _getWidth2 = _interopRequireDefault(_getWidth);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Radio = function (_ref) {
  (0, _inherits3['default'])(Radio, _ref);

  function Radio() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Radio);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = Radio.__proto__ || Object.getPrototypeOf(Radio)).call.apply(_ref2, [this].concat(args))), _this), _this.handleChange = function (evt) {
      var props = _this.props;

      props.onChange({
        target: (0, _extends3['default'])({}, props, {
          type: 'radio',
          checked: evt.target.checked
        }),

        preventDefault: function preventDefault() {
          evt.preventDefault();
        },
        stopPropagation: function stopPropagation() {
          evt.stopPropagation();
        }
      });
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  // event liftup
  // link: https://facebook.github.io/react/docs/lifting-state-up.html


  (0, _createClass3['default'])(Radio, [{
    key: 'render',
    value: function render() {
      var _classNames;

      var _props = this.props,
          checked = _props.checked,
          className = _props.className,
          style = _props.style,
          prefix = _props.prefix,
          disabled = _props.disabled,
          readOnly = _props.readOnly,
          children = _props.children,
          value = _props.value,
          width = _props.width,
          others = (0, _objectWithoutProperties3['default'])(_props, ['checked', 'className', 'style', 'prefix', 'disabled', 'readOnly', 'children', 'value', 'width']);


      var classString = (0, _classnames2['default'])((_classNames = {}, (0, _defineProperty3['default'])(_classNames, className, !!className), (0, _defineProperty3['default'])(_classNames, prefix + '-radio-wrap', true), (0, _defineProperty3['default'])(_classNames, prefix + '-radio-checked', !!checked), (0, _defineProperty3['default'])(_classNames, prefix + '-radio-disabled', disabled || readOnly), _classNames));
      var widthStyle = (0, _getWidth2['default'])(width);
      var wrapStyle = (0, _assign2['default'])({}, style, widthStyle);
      return _react2['default'].createElement(
        'label',
        { className: classString, style: wrapStyle },
        _react2['default'].createElement(
          'span',
          { className: prefix + '-radio' },
          _react2['default'].createElement('span', { className: prefix + '-radio-inner' }),
          _react2['default'].createElement('input', (0, _extends3['default'])({}, others, {
            type: 'radio',
            checked: !!checked,
            disabled: disabled,
            readOnly: readOnly,
            onChange: this.handleChange
          }))
        ),
        children !== undefined ? _react2['default'].createElement(
          'span',
          null,
          children
        ) : null
      );
    }
  }]);
  return Radio;
}(_react.PureComponent || _react.Component);

Radio.propTypes = {
  checked: _propTypes2['default'].oneOfType([_propTypes2['default'].number, _propTypes2['default'].bool]),
  value: _propTypes2['default'].any,
  disabled: _propTypes2['default'].bool,
  readOnly: _propTypes2['default'].bool,

  // will be rewrite by RadioGroup
  onChange: _propTypes2['default'].func,
  className: _propTypes2['default'].string,
  style: _propTypes2['default'].object,
  prefix: _propTypes2['default'].string,
  width: _propTypes2['default'].oneOfType([_propTypes2['default'].number, _propTypes2['default'].string])
};
Radio.defaultProps = {
  prefix: 'zent',
  className: '',
  style: {},
  onChange: function onChange() {}
};
exports['default'] = Radio;
