// zent/lib/pagination/modules/CorePagination.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _parser = require('../data/parser');

var _parser2 = _interopRequireDefault(_parser);

var _Pager = require('./Pager');

var _Pager2 = _interopRequireDefault(_Pager);

var _Jump = require('./Jump');

var _Jump2 = _interopRequireDefault(_Jump);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var number = _propTypes2['default'].number,
    func = _propTypes2['default'].func;

var CorePagination = function (_ref) {
  (0, _inherits3['default'])(CorePagination, _ref);

  function CorePagination() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, CorePagination);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = CorePagination.__proto__ || Object.getPrototypeOf(CorePagination)).call.apply(_ref2, [this].concat(args))), _this), _this.renderPager = function (item, index) {
      if (!item) {
        return null;
      }

      if (item.type === 'input') {
        return _react2['default'].createElement(_Jump2['default'], {
          key: index,
          content: item.content,
          total: item.total,
          onChange: _this.props.onChange
        });
      }

      return _react2['default'].createElement(_Pager2['default'], {
        key: index,
        content: item.content,
        current: item.current,
        target: item.target,
        type: item.type,
        onChange: _this.props.onChange
      });
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(CorePagination, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          current = _props.current,
          total = _props.total,
          maxPageToShow = _props.maxPageToShow;


      if (maxPageToShow && maxPageToShow > 0) {
        total = Math.min(total, maxPageToShow);
      }

      var pages = _parser2['default'].getPages({
        current: current,
        total: total
      });

      return _react2['default'].createElement(
        'div',
        { className: 'pagination-list' },
        pages.map(function (item, index) {
          return _this2.renderPager(item, index);
        })
      );
    }
  }]);
  return CorePagination;
}(_react.PureComponent || _react.Component);

CorePagination.propTypes = {
  current: number,
  total: number,
  onChange: func
};
exports['default'] = CorePagination;
