// zent/lib/pagination/modules/Jump.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _i18n = require('../../i18n');

var _default = require('../../i18n/default');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Jump = function (_ref) {
  (0, _inherits3['default'])(Jump, _ref);

  function Jump() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Jump);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = Jump.__proto__ || Object.getPrototypeOf(Jump)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      pageLabel: _this.props.content.trim()
    }, _this.onKeyUp = function (e) {
      if (e.key !== 'Enter') return;
      var value = e.target.value.trim();
      var pattern = /^\d+$/g;

      if (pattern.test(value)) {
        if (value <= 0) {
          value = 0;
        }

        if (value > _this.props.total) {
          value = _this.props.total;
        }

        _this.props.onChange(parseInt(value, 10));
      }
    }, _this.onChange = function (e) {
      _this.setState({
        pageLabel: e.target.value.trim()
      });
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Jump, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.setState({
        pageLabel: nextProps.content.trim()
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var total = this.props.total;

      return _react2['default'].createElement(
        'div',
        { className: 'pager pager--jump' },
        _react2['default'].createElement('input', {
          className: 'pager__input',
          value: this.state.pageLabel,
          onKeyUp: this.onKeyUp,
          onChange: this.onChange
        }),
        _react2['default'].createElement(
          _i18n.I18nReceiver,
          { componentName: 'Pagination', defaultI18n: _default.Pagination },
          function (i18n) {
            return _react2['default'].createElement(
              'span',
              { className: 'pager__suffix' },
              '/ ' + total + ' ' + i18n.pages
            );
          }
        )
      );
    }
  }]);
  return Jump;
}(_react.PureComponent || _react.Component);

exports['default'] = Jump;
