// zent/lib/pagination/modules/Pager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var string = _propTypes2['default'].string,
    number = _propTypes2['default'].number,
    func = _propTypes2['default'].func,
    bool = _propTypes2['default'].bool;

var Pager = function (_ref) {
  (0, _inherits3['default'])(Pager, _ref);

  function Pager() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Pager);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = Pager.__proto__ || Object.getPrototypeOf(Pager)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      pageLabel: _this.props.content
    }, _this.onClick = function () {
      var _this$props = _this.props,
          target = _this$props.target,
          current = _this$props.current,
          onChange = _this$props.onChange;

      // 有目标，且不是当前页的时候

      if (target && !current) {
        onChange(target);
      }
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Pager, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.setState({
        pageLabel: nextProps.content
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          current = _props.current,
          type = _props.type,
          target = _props.target;

      var className = (0, _classnames2['default'])({
        pager: true,
        'pager--current': current,
        'pager--omni': type === 'omni',
        'pager--disabled': type !== 'omni' && !target
      });

      return _react2['default'].createElement(
        'div',
        { className: className, onClick: this.onClick },
        this.props.content
      );
    }
  }]);
  return Pager;
}(_react.PureComponent || _react.Component);

Pager.propTypes = {
  content: string,
  current: bool,
  onClick: func,
  target: number
};
exports['default'] = Pager;
