// zent/lib/pagination/modules/Prefix.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _select = require('../../select');

var _select2 = _interopRequireDefault(_select);

var _i18n = require('../../i18n');

var _default = require('../../i18n/default');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var number = _propTypes2['default'].number,
    array = _propTypes2['default'].array,
    oneOfType = _propTypes2['default'].oneOfType,
    func = _propTypes2['default'].func;
var Option = _select2['default'].Option;

var Prefix = function (_ref) {
  (0, _inherits3['default'])(Prefix, _ref);

  function Prefix() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Prefix);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = Prefix.__proto__ || Object.getPrototypeOf(Prefix)).call.apply(_ref2, [this].concat(args))), _this), _this.changePageSize = function (e, data) {
      _this.props.setPageSize(data.text);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Prefix, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          totalItem = _props.totalItem,
          pageSize = _props.pageSize,
          currentPageSize = _props.currentPageSize;

      var isNeedSelect = Array.isArray(pageSize) && pageSize.length > 1;
      var pageSizeForSelect = pageSize.map(function (item) {
        return '' + item.value;
      });
      return _react2['default'].createElement(
        'span',
        { className: 'zent-pagination__info' },
        _react2['default'].createElement(
          _i18n.I18nReceiver,
          { componentName: 'Pagination', defaultI18n: _default.Pagination },
          function (i18n) {
            return _react2['default'].createElement(
              'span',
              { className: 'total each' },
              i18n.total + ' ' + totalItem + ' ' + i18n.items + i18n.comma,
              isNeedSelect ? _react2['default'].createElement(
                _select2['default'],
                {
                  value: currentPageSize,
                  onChange: _this2.changePageSize,
                  width: 60,
                  autoWidth: true
                },
                pageSizeForSelect.map(function (item, i) {
                  return _react2['default'].createElement(
                    Option,
                    { key: i, value: item },
                    item
                  );
                })
              ) : currentPageSize,
              '' + i18n.items + i18n.perPage
            );
          }
        )
      );
    }
  }]);
  return Prefix;
}(_react.PureComponent || _react.Component);

Prefix.propTypes = {
  totalItem: number,
  pageSize: oneOfType([number, array]),
  setPageSize: func
};
exports['default'] = Prefix;
