// zent/lib/menu/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _CommonMenu2 = require('./CommonMenu');

var _CommonMenu3 = _interopRequireDefault(_CommonMenu2);

var _MenuItem = require('./MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _SubMenu = require('./SubMenu');

var _SubMenu2 = _interopRequireDefault(_SubMenu);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Menu = function (_CommonMenu) {
  (0, _inherits3['default'])(Menu, _CommonMenu);

  function Menu() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, Menu);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref = Menu.__proto__ || Object.getPrototypeOf(Menu)).call.apply(_ref, [this].concat(args))), _this), _this.handleClick = function (e, key) {
      var onClick = _this.props.onClick;

      onClick && onClick(e, key);
    }, _this.renderMenuItem = function (c, i, index) {
      if (!c) {
        return null;
      }

      return _this.renderCommonMenuItem(c, i, index);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(Menu, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          prefix = _props.prefix,
          className = _props.className,
          style = _props.style;


      return _react2['default'].createElement(
        'ul',
        { className: (0, _classnames2['default'])(prefix + '-menu', className), style: style },
        _react2['default'].Children.map(children, this.renderMenuItem)
      );
    }
  }]);
  return Menu;
}(_CommonMenu3['default']);

Menu.MenuItem = _MenuItem2['default'];
Menu.SubMenu = _SubMenu2['default'];
Menu.propTypes = {
  onClick: _propTypes2['default'].func,
  style: _propTypes2['default'].object,
  className: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string
};
Menu.defaultProps = {
  prefix: 'zent',
  onClick: _noop2['default']
};
exports['default'] = Menu;
