// zent/lib/menu/SubMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _icon = require('../icon');

var _icon2 = _interopRequireDefault(_icon);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _SubPopupMenu = require('./SubPopupMenu');

var _SubPopupMenu2 = _interopRequireDefault(_SubPopupMenu);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var SubMenu = function (_ref) {
  (0, _inherits3['default'])(SubMenu, _ref);

  function SubMenu() {
    var _ref2;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, SubMenu);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref2 = SubMenu.__proto__ || Object.getPrototypeOf(SubMenu)).call.apply(_ref2, [this].concat(args))), _this), _this.state = {
      subMenuVisible: false
    }, _this.handleClick = function (e, index) {
      var onClick = _this.props.onClick;

      _this.setState({ subMenuVisible: false });
      onClick(e, index);
    }, _this.titleClickHandler = function (e) {
      e.stopPropagation();
    }, _this.onMouseEnter = function () {
      if (_this.leaveTimer) {
        clearTimeout(_this.leaveTimer);
      }
      _this.enterTimer = setTimeout(function () {
        _this.setState({ subMenuVisible: true });
      }, 200);
    }, _this.onMouseLeave = function () {
      if (_this.enterTimer) {
        clearTimeout(_this.enterTimer);
      }
      _this.leaveTimer = setTimeout(function () {
        _this.setState({ subMenuVisible: false });
      }, 200);
    }, _this.renderContent = function () {
      var _this$props = _this.props,
          prefix = _this$props.prefix,
          children = _this$props.children,
          index = _this$props.index,
          overlayClassName = _this$props.overlayClassName;


      return _react2['default'].createElement(
        _SubPopupMenu2['default'],
        {
          prefix: prefix,
          visible: _this.state.subMenuVisible,
          onClick: _this.handleClick,
          index: index,
          overlayCx: overlayClassName
        },
        children
      );
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(SubMenu, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          prefix = _props.prefix,
          className = _props.className,
          disabled = _props.disabled,
          title = _props.title;


      var mouseEvents = disabled ? {} : {
        onMouseEnter: this.onMouseEnter,
        onMouseLeave: this.onMouseLeave
      };
      var cls = (0, _classnames2['default'])(prefix + '-menu-item', prefix + '-submenu', className, (0, _defineProperty3['default'])({}, prefix + '-menu-item-disabled', disabled));

      return _react2['default'].createElement(
        'li',
        (0, _extends3['default'])({ className: cls }, mouseEvents),
        _react2['default'].createElement(
          'div',
          {
            className: (0, _classnames2['default'])(prefix + '-submenu-title', (0, _defineProperty3['default'])({}, prefix + '-submenu-disabled', disabled)),
            onClick: this.titleClickHandler
          },
          title,
          !disabled && _react2['default'].createElement(_icon2['default'], { type: 'right' })
        ),
        !disabled && this.renderContent()
      );
    }
  }]);
  return SubMenu;
}(_react.PureComponent || _react.Component);

SubMenu.propTypes = {
  title: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string,
  className: _propTypes2['default'].string,
  overlayClassName: _propTypes2['default'].string,
  disabled: _propTypes2['default'].bool,
  onClick: _propTypes2['default'].func
};
SubMenu.defaultProps = {
  className: '',
  prefix: 'zent'
};
exports['default'] = SubMenu;
