// zent/lib/menu/SubPopupMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _CommonMenu2 = require('./CommonMenu');

var _CommonMenu3 = _interopRequireDefault(_CommonMenu2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var PopupMenu = function (_CommonMenu) {
  (0, _inherits3['default'])(PopupMenu, _CommonMenu);

  function PopupMenu() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, PopupMenu);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref = PopupMenu.__proto__ || Object.getPrototypeOf(PopupMenu)).call.apply(_ref, [this].concat(args))), _this), _this.handleClick = function (e, index) {
      var onClick = _this.props.onClick;

      onClick(e, index);
    }, _this.renderSubMenuItems = function (c, i) {
      if (!c) {
        return null;
      }
      var index = _this.props.index;

      return _this.renderCommonMenuItem(c, i, index);
    }, _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(PopupMenu, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          prefix = _props.prefix,
          visible = _props.visible,
          overlayCx = _props.overlayCx;

      if (!visible) {
        return null;
      }

      return _react2['default'].createElement(
        'ul',
        {
          className: (0, _classnames2['default'])(prefix + '-menu', prefix + '-submenu-content', overlayCx)
        },
        _react2['default'].Children.map(children, this.renderSubMenuItems)
      );
    }
  }]);
  return PopupMenu;
}(_CommonMenu3['default']);

PopupMenu.propTypes = {
  prefix: _propTypes2['default'].string,
  visible: _propTypes2['default'].bool,
  onClick: _propTypes2['default'].func,
  index: _propTypes2['default'].string,
  overlayCx: _propTypes2['default'].string
};
exports['default'] = PopupMenu;
