// zent/lib/design/components/line/LinePreview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var LinePreview = function (_ref) {
  (0, _inherits3['default'])(LinePreview, _ref);

  function LinePreview() {
    (0, _classCallCheck3['default'])(this, LinePreview);
    return (0, _possibleConstructorReturn3['default'])(this, (LinePreview.__proto__ || Object.getPrototypeOf(LinePreview)).apply(this, arguments));
  }

  (0, _createClass3['default'])(LinePreview, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          value = _props.value,
          prefix = _props.prefix;


      return _react2['default'].createElement(
        'div',
        { className: prefix + '-design-component-line-preview' },
        _react2['default'].createElement('div', { style: createStyle(value) })
      );
    }
  }]);
  return LinePreview;
}(_react.PureComponent || _react.Component);

exports['default'] = LinePreview;


function createStyle(value) {
  var color = value.color,
      hasPadding = value.hasPadding,
      lineType = value.lineType;


  return {
    height: 0,
    borderTopWidth: '1px',
    margin: hasPadding ? '0 10px' : 0,
    borderColor: color,
    borderStyle: lineType
  };
}
