// zent/lib/datetimepicker/common/PanelFooter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _button = require('../../button');

var _button2 = _interopRequireDefault(_button);

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var PanelFooter = function PanelFooter(props) {
  var buttonText = props.buttonText,
      errorText = props.errorText,
      linkCls = props.linkCls,
      linkText = props.linkText,
      onClickButton = props.onClickButton,
      onClickLink = props.onClickLink,
      showError = props.showError,
      showLink = props.showLink;


  return _react2['default'].createElement(
    'div',
    { className: 'panel__footer' },
    showLink && _react2['default'].createElement(
      'a',
      { className: linkCls, onClick: onClickLink },
      linkText
    ),
    showError && _react2['default'].createElement(
      'span',
      { className: 'error-tips' },
      errorText
    ),
    _react2['default'].createElement(
      _button2['default'],
      { className: 'btn--confirm', type: 'primary', onClick: onClickButton },
      buttonText
    )
  );
};

PanelFooter.defaultProps = {
  onClickLink: _constants.noop,
  onClickButton: _constants.noop,
  showLink: true,
  showError: false,
  errorText: ''
};

exports['default'] = PanelFooter;
