// zent/lib/table/helper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _intersection = require('lodash/intersection');

var _intersection2 = _interopRequireDefault(_intersection);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var helper = {
  getCalculatedWidth: function getCalculatedWidth(width) {
    var res = void 0;
    if (typeof width === 'number') {
      res = width + '%';
    } else if (typeof width === 'string') {
      res = width;
    }

    return res;
  },
  isReactComponent: function isReactComponent(render) {
    var isReact = false;

    if (typeof render === 'function') {
      if (render.prototype && render.prototype.isReactComponent) {
        isReact = true;
      }
    }

    return isReact;
  },
  getAlignClass: function getAlignClass(textAlign) {
    var alignValue = '';

    if (textAlign) {
      switch (textAlign) {
        case 'left':
          alignValue = 'start';
          break;
        case 'right':
          alignValue = 'end';
          break;
        case 'center':
          alignValue = 'center';
          break;
        default:
          alignValue = 'start';
      }
    }

    return alignValue;
  },


  // return true/false/null
  needFixBatchComps: function needFixBatchComps(isTableInView, isFootInView, isSelectRows, isFixed) {
    var res = null;
    if (isTableInView && !isFootInView && isSelectRows) {
      if (!isFixed) {
        res = true;
      }
    } else if (isFixed) {
      res = false;
    }

    return res;
  },


  /**
   * 是否选中全部
   * @param {array} rowKeys
   * @param {array} subRowKeys
   * @return {boolean} isSelectAll
   */
  isSelectAll: function isSelectAll(rowKeys, subRowKeys) {
    return (0, _intersection2['default'])(rowKeys, subRowKeys).length === subRowKeys.length;
  },


  /**
   * 是否选中了部分
   * @param {array} rowKeys
   * @param {array} subRowKeys
   * @return {boolean} 是否选中了部分
   */
  isSelectPart: function isSelectPart(rowKeys, subRowKeys) {
    return (0, _intersection2['default'])(rowKeys, subRowKeys).length > 0;
  },


  /**
   * 切换是否监听事件
   */
  toggleEventListener: function toggleEventListener(currentProps, nextProps) {
    var toggleEventListener = void 0;
    if (nextProps.batchComponentsAutoFixed !== currentProps.batchComponentsAutoFixed) {
      if (nextProps.batchComponentsAutoFixed) {
        toggleEventListener = 'addEventListener';
      } else {
        toggleEventListener = 'removeEventListener';
      }
    }

    return toggleEventListener;
  }
};

exports['default'] = helper;
