// zent/lib/steps/Steps.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _NumberSteps = require('./components/NumberSteps');

var _NumberSteps2 = _interopRequireDefault(_NumberSteps);

var _BreadcrumbSteps = require('./components/BreadcrumbSteps');

var _BreadcrumbSteps2 = _interopRequireDefault(_BreadcrumbSteps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var Steps = function (_ref) {
  (0, _inherits3['default'])(Steps, _ref);

  function Steps() {
    (0, _classCallCheck3['default'])(this, Steps);
    return (0, _possibleConstructorReturn3['default'])(this, (Steps.__proto__ || Object.getPrototypeOf(Steps)).apply(this, arguments));
  }

  (0, _createClass3['default'])(Steps, [{
    key: 'render',
    value: function render() {
      var props = this.props;
      var type = props.type,
          children = props.children,
          restProps = (0, _objectWithoutProperties3['default'])(props, ['type', 'children']);

      var typeComponentMapping = {
        number: _NumberSteps2['default'],
        card: _BreadcrumbSteps2['default'],
        breadcrumb: _BreadcrumbSteps2['default']
      };

      var StepsComponent = typeComponentMapping[type];

      return _react2['default'].createElement(
        StepsComponent,
        (0, _extends3['default'])({}, restProps, { type: type }),
        children
      );
    }
  }]);
  return Steps;
}(_react.PureComponent || _react.Component);

Steps.propTypes = {
  className: _propTypes2['default'].string,
  prefix: _propTypes2['default'].string,
  children: _propTypes2['default'].any,
  type: _propTypes2['default'].string,
  current: _propTypes2['default'].number,
  direction: _propTypes2['default'].string,
  size: _propTypes2['default'].string,
  status: _propTypes2['default'].string,
  sequence: _propTypes2['default'].bool,
  onStepChange: _propTypes2['default'].func
};
Steps.defaultProps = {
  className: '',
  prefix: 'zent',
  type: 'number',
  current: 0,
  direction: 'horizontal',
  size: 'normal',
  status: 'finish',
  sequence: true
};
exports['default'] = Steps;
