// zent/lib/design/components/whitespace/WhitespaceEditor.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _slider = require('../../../slider');

var _slider2 = _interopRequireDefault(_slider);

var _DesignEditor2 = require('../../editor/DesignEditor');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var WhitespaceEditor = function (_DesignEditor) {
  (0, _inherits3['default'])(WhitespaceEditor, _DesignEditor);

  function WhitespaceEditor() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3['default'])(this, WhitespaceEditor);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3['default'])(this, (_ref = WhitespaceEditor.__proto__ || Object.getPrototypeOf(WhitespaceEditor)).call.apply(_ref, [this].concat(args))), _this), _this.onHeightChange = _this.onCustomInputChange('height'), _temp), (0, _possibleConstructorReturn3['default'])(_this, _ret);
  }

  (0, _createClass3['default'])(WhitespaceEditor, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          value = _props.value,
          prefix = _props.prefix;


      return _react2['default'].createElement(
        'div',
        { className: prefix + '-design-component-whitespace-editor' },
        _react2['default'].createElement(
          _DesignEditor2.ControlGroup,
          {
            label: '\u7A7A\u767D\u9AD8\u5EA6:',
            className: prefix + '-design-component-whitespace-editor__height'
          },
          _react2['default'].createElement(_slider2['default'], {
            min: 10,
            max: 100,
            value: value.height,
            onChange: this.onHeightChange,
            withInput: false
          }),
          _react2['default'].createElement(
            'span',
            null,
            value.height,
            ' \u50CF\u7D20'
          )
        )
      );
    }

    // 组件的类型


    // 组件的描述

  }], [{
    key: 'getInitialValue',


    // 添加组件时调用，用来获取新组件的初始值
    value: function getInitialValue() {
      return {
        height: 30
      };
    }
  }]);
  return WhitespaceEditor;
}(_DesignEditor2.DesignEditor);

WhitespaceEditor.designType = 'white';
WhitespaceEditor.designDescription = _react2['default'].createElement(
  'span',
  null,
  '\u8F85\u52A9',
  _react2['default'].createElement('br', null),
  '\u7A7A\u767D'
);
exports['default'] = WhitespaceEditor;
